From b35615d0257a50e7f1f8e18ccb22b16e6d558627 Mon Sep 17 00:00:00 2001
From: bigfoot547 <bigfoot@figboot.dev>
Date: Thu, 10 Mar 2022 01:37:59 -0600
Subject: [PATCH] Add maven project


diff --git a/.gitignore b/.gitignore
new file mode 100644
index 0000000..5231862
--- /dev/null
+++ b/.gitignore
@@ -0,0 +1,3 @@
+.idea
+*.iml
+target
diff --git a/Log4j-config.xsd b/Log4j-config.xsd
deleted file mode 100644
index a19e125..0000000
--- a/Log4j-config.xsd
+++ /dev/null
@@ -1,133 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
-<!--
- Licensed to the Apache Software Foundation (ASF) under one or more
- contributor license agreements.  See the NOTICE file distributed with
- this work for additional information regarding copyright ownership.
- The ASF licenses this file to You under the Apache License, Version 2.0
- (the "License"); you may not use this file except in compliance with
- the License.  You may obtain a copy of the License at
-
-      http://www.apache.org/licenses/LICENSE-2.0
-
- Unless required by applicable law or agreed to in writing, software
- distributed under the License is distributed on an "AS IS" BASIS,
- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- See the License for the specific language governing permissions and
- limitations under the License.
-
--->
-<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified" attributeFormDefault="unqualified">
-    <xs:element name="Configuration" type="ConfigurationType"/>
-    <xs:complexType name="ConfigurationType">
-        <xs:sequence>
-            <xs:element name="Properties" type="PropertiesType"/>
-            <xs:choice minOccurs="0" maxOccurs="1">
-                <xs:element name="Filters" type="FiltersType"/>
-                <xs:element name="Filter" type="FilterType"/>
-            </xs:choice>
-            <xs:element name="ThresholdFilter" type="ThresholdFilterType"/>
-            <xs:element name="Appenders" type="AppendersType"/>
-            <xs:element name="Loggers" type="LoggersType"/>
-        </xs:sequence>
-        <xs:attribute name="packages" type="xs:string"/>
-        <xs:attribute name="status" type="xs:string"/>
-        <xs:attribute name="strict" type="xs:string"/>
-        <xs:attribute name="name" type="xs:string"/>
-        <xs:attribute name="advertiser" type="xs:string"/>
-    </xs:complexType>
-    <xs:complexType name="PropertiesType">
-        <xs:sequence>
-            <xs:element name="Property" type="PropertyType"/>
-        </xs:sequence>
-    </xs:complexType>
-    <xs:complexType name="AppenderType">
-        <xs:sequence>
-            <xs:element name="Layout" type="LayoutType" minOccurs="0"/>
-            <xs:choice minOccurs="0" maxOccurs="1">
-                <xs:element name="Filters" type="FiltersType"/>
-                <xs:element name="Filter" type="FilterType"/>
-            </xs:choice>
-        </xs:sequence>
-        <xs:attribute name="type" type="xs:string" use="required"/>
-        <xs:attribute name="name" type="xs:string" use="required"/>
-        <xs:attribute name="fileName" type="xs:string" use="optional"/>
-    </xs:complexType>
-    <xs:complexType name="RootType">
-        <xs:sequence>
-            <xs:element name="AppenderRef" type="AppenderRefType" minOccurs="1" maxOccurs="unbounded"/>
-        </xs:sequence>
-        <xs:attribute name="level" type="xs:string"/>
-    </xs:complexType>
-    <xs:complexType name="PropertyType">
-        <xs:simpleContent>
-            <xs:extension base="xs:string">
-                <xs:attribute name="name" type="xs:string"/>
-            </xs:extension>
-        </xs:simpleContent>
-    </xs:complexType>
-    <xs:complexType name="KeyValuePairType">
-        <xs:simpleContent>
-            <xs:extension base="xs:string">
-                <xs:attribute name="key" type="xs:string"/>
-                <xs:attribute name="value" type="xs:string"/>
-            </xs:extension>
-        </xs:simpleContent>
-    </xs:complexType>
-    <xs:complexType name="AppendersType">
-        <xs:sequence>
-            <xs:element name="Appender" type="AppenderType" minOccurs="1" maxOccurs="unbounded"/>
-        </xs:sequence>
-    </xs:complexType>
-    <xs:complexType name="AppenderRefType">
-        <xs:simpleContent>
-            <xs:extension base="xs:string">
-                <xs:attribute name="ref" type="xs:string" use="required"/>
-            </xs:extension>
-        </xs:simpleContent>
-    </xs:complexType>
-    <xs:complexType name="LoggerType">
-        <xs:sequence>
-            <xs:choice minOccurs="0" maxOccurs="1">
-                <xs:element name="Filters" type="FiltersType"/>
-                <xs:element name="Filter" type="FilterType"/>
-            </xs:choice>
-            <xs:element name="AppenderRef" type="AppenderRefType"/>
-        </xs:sequence>
-        <xs:attribute name="name" type="xs:string" use="required"/>
-        <xs:attribute name="level" type="xs:string" use="optional"/>
-        <xs:attribute name="additivity" type="xs:string" use="optional"/>
-    </xs:complexType>
-    <xs:complexType name="FilterType" mixed="true">
-        <xs:sequence>
-            <xs:element name="KeyValuePair" type="KeyValuePairType" minOccurs="0"/>
-        </xs:sequence>
-        <xs:attribute name="type" type="xs:string" use="required"/>
-        <xs:attribute name="level" type="xs:string" use="optional"/>
-        <xs:attribute name="marker" type="xs:string" use="optional"/>
-        <xs:attribute name="onMatch" type="xs:string" use="optional"/>
-        <xs:attribute name="onMismatch" type="xs:string" use="optional"/>
-    </xs:complexType>
-    <xs:complexType name="FiltersType">
-        <xs:sequence>
-            <xs:element name="Filter" type="FilterType" minOccurs="0" maxOccurs="unbounded"/>
-        </xs:sequence>
-    </xs:complexType>
-    <xs:complexType name="LoggersType" mixed="true">
-        <xs:sequence>
-            <xs:element name="Logger" type="LoggerType" minOccurs="0" maxOccurs="unbounded"/>
-            <xs:element name="Root" type="RootType" minOccurs="1" maxOccurs="1"/>
-        </xs:sequence>
-    </xs:complexType>
-    <xs:complexType name="LayoutType" mixed="true">
-        <xs:sequence>
-            <xs:element name="Pattern" type="xs:string" minOccurs="0"/>
-        </xs:sequence>
-        <xs:attribute name="Type" type="xs:string" use="required"/>
-        <xs:attribute name="Pattern" type="xs:string" use="optional"/>
-    </xs:complexType>
-    <xs:complexType name="ThresholdFilterType">
-        <xs:attribute name="level" type="xs:string" use="optional"/>
-        <xs:attribute name="onMatch" type="xs:string" use="optional"/>
-        <xs:attribute name="onMismatch" type="xs:string" use="optional"/>
-    </xs:complexType>
-</xs:schema>
diff --git a/Log4j-events.dtd b/Log4j-events.dtd
deleted file mode 100644
index 91af8c7..0000000
--- a/Log4j-events.dtd
+++ /dev/null
@@ -1,64 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
-<!--
- Licensed to the Apache Software Foundation (ASF) under one or more
- contributor license agreements.  See the NOTICE file distributed with
- this work for additional information regarding copyright ownership.
- The ASF licenses this file to You under the Apache License, Version 2.0
- (the "License"); you may not use this file except in compliance with
- the License.  You may obtain a copy of the License at
-
-      http://www.apache.org/licenses/LICENSE-2.0
-
- Unless required by applicable law or agreed to in writing, software
- distributed under the License is distributed on an "AS IS" BASIS,
- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- See the License for the specific language governing permissions and
- limitations under the License.
--->
-<!--DTD generated by XMLSpy v2011 rel. 3 sp1 (x64) (http://www.altova.com)-->
-<!--the entity declarations may be overridden in the internal subset-->
-<!--namespace prefixes-->
-<!ENTITY % log4j_prefix "log4j">
-<!--namespace prefix to namespace uri mappings-->
-<!ENTITY % log4j_prefix.. "%log4j_prefix;:">
-<!--namespaces attributes for root element-->
-<!ENTITY % documentElementAttributes " xmlns:%log4j_prefix; CDATA 'http://logging.apache.org/log4j/2.0/events'">
-<!--the declarations below should not be modified-->
-<!--element name mappings-->
-<!ENTITY % log4j..Events "%log4j_prefix..;Events">
-<!ENTITY % log4j..Event "%log4j_prefix..;Event">
-<!ENTITY % log4j..Message "%log4j_prefix..;Message">
-<!ENTITY % log4j..NDC "%log4j_prefix..;NDC">
-<!ENTITY % log4j..Throwable "%log4j_prefix..;Throwable">
-<!ENTITY % log4j..LocationInfo "%log4j_prefix..;LocationInfo">
-<!ENTITY % log4j..Properties "%log4j_prefix..;Properties">
-<!ENTITY % log4j..Data "%log4j_prefix..;Data">
-<!--element and attribute declarations-->
-<!--Log4J 2.0 XML Schema-->
-<!ELEMENT %log4j..Events; ((%log4j..Event;)*)>
-<!ATTLIST %log4j..Events;
-	%documentElementAttributes;
->
-<!ELEMENT %log4j..Event; (%log4j..Message;, (%log4j..NDC;)?, (%log4j..Throwable;)?, (%log4j..LocationInfo;)?, (%log4j..Properties;)?)>
-<!ATTLIST %log4j..Event;
-	logger CDATA #REQUIRED
-	timestamp NMTOKEN #REQUIRED
-	level (OFF | FATAL | ERROR | WARN | INFO | DEBUG | TRACE | ALL) #REQUIRED
-	thread CDATA #REQUIRED
->
-<!ELEMENT %log4j..Message; ANY>
-<!ELEMENT %log4j..NDC; ANY>
-<!ELEMENT %log4j..Throwable; ANY>
-<!ELEMENT %log4j..LocationInfo; EMPTY>
-<!ATTLIST %log4j..LocationInfo;
-	class CDATA #REQUIRED
-	method CDATA #REQUIRED
-	file CDATA #REQUIRED
-	line NMTOKEN #REQUIRED
->
-<!ELEMENT %log4j..Properties; ((%log4j..Data;)+)>
-<!ELEMENT %log4j..Data; EMPTY>
-<!ATTLIST %log4j..Data;
-	name CDATA #REQUIRED
-	value CDATA #REQUIRED
->
diff --git a/Log4j-events.xsd b/Log4j-events.xsd
deleted file mode 100644
index 2310227..0000000
--- a/Log4j-events.xsd
+++ /dev/null
@@ -1,74 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
-<!-- edited with XMLSpy v2011 rel. 3 sp1 (x64) (http://www.altova.com) by Gary Gregory (Rocket Software) -->
-<!--
- Licensed to the Apache Software Foundation (ASF) under one or more
- contributor license agreements.  See the NOTICE file distributed with
- this work for additional information regarding copyright ownership.
- The ASF licenses this file to You under the Apache License, Version 2.0
- (the "License"); you may not use this file except in compliance with
- the License.  You may obtain a copy of the License at
-
-      http://www.apache.org/licenses/LICENSE-2.0
-
- Unless required by applicable law or agreed to in writing, software
- distributed under the License is distributed on an "AS IS" BASIS,
- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- See the License for the specific language governing permissions and
- limitations under the License.
--->
-<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:log4j="http://logging.apache.org/log4j/2.0/events" targetNamespace="http://logging.apache.org/log4j/2.0/events" elementFormDefault="qualified" attributeFormDefault="unqualified">
-    <xs:element name="Events">
-        <xs:annotation>
-            <xs:documentation>Log4J 2.0 XML Schema for XML log event files.</xs:documentation>
-        </xs:annotation>
-        <xs:complexType>
-            <xs:sequence>
-                <xs:element name="Event" minOccurs="0" maxOccurs="unbounded">
-                    <xs:complexType>
-                        <xs:sequence>
-                            <xs:element name="Message"/>
-                            <xs:element name="NDC" minOccurs="0"/>
-                            <xs:element name="Throwable" minOccurs="0"/>
-                            <xs:element name="LocationInfo" minOccurs="0">
-                                <xs:complexType>
-                                    <xs:attribute name="class" type="xs:string" use="required"/>
-                                    <xs:attribute name="method" type="xs:string" use="required"/>
-                                    <xs:attribute name="file" type="xs:string" use="required"/>
-                                    <xs:attribute name="line" type="xs:int" use="required"/>
-                                </xs:complexType>
-                            </xs:element>
-                            <xs:element name="Properties" minOccurs="0">
-                                <xs:complexType>
-                                    <xs:sequence>
-                                        <xs:element name="Data" maxOccurs="unbounded">
-                                            <xs:complexType>
-                                                <xs:attribute name="name" type="xs:string" use="required"/>
-                                                <xs:attribute name="value" use="required"/>
-                                            </xs:complexType>
-                                        </xs:element>
-                                    </xs:sequence>
-                                </xs:complexType>
-                            </xs:element>
-                        </xs:sequence>
-                        <xs:attribute name="logger" type="xs:string" use="required"/>
-                        <xs:attribute name="timestamp" type="xs:long" use="required"/>
-                        <xs:attribute name="level" type="log4j:LevelEnum" use="required"/>
-                        <xs:attribute name="thread" type="xs:string" use="required"/>
-                    </xs:complexType>
-                </xs:element>
-            </xs:sequence>
-        </xs:complexType>
-    </xs:element>
-    <xs:simpleType name="LevelEnum">
-        <xs:restriction base="xs:string">
-            <xs:enumeration value="OFF"/>
-            <xs:enumeration value="FATAL"/>
-            <xs:enumeration value="ERROR"/>
-            <xs:enumeration value="WARN"/>
-            <xs:enumeration value="INFO"/>
-            <xs:enumeration value="DEBUG"/>
-            <xs:enumeration value="TRACE"/>
-            <xs:enumeration value="ALL"/>
-        </xs:restriction>
-    </xs:simpleType>
-</xs:schema>
diff --git a/META-INF/MANIFEST.MF b/META-INF/MANIFEST.MF
deleted file mode 100644
index aef7cf0..0000000
--- a/META-INF/MANIFEST.MF
+++ /dev/null
@@ -1,4 +0,0 @@
-Manifest-Version: 1.0
-Implementation-Version: 1.6.93
-Main-Class: net.minecraft.launcher.Main
-
diff --git a/pom.xml b/pom.xml
new file mode 100644
index 0000000..b8a1b9a
--- /dev/null
+++ b/pom.xml
@@ -0,0 +1,428 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>dev.figboot</groupId>
+        <artifactId>olauncher-parent</artifactId>
+        <version>0.1.0</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>olauncher</artifactId>
+    <version>1.7.2_02</version>
+
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <lwjgl.version>3.3.1</lwjgl.version>
+    </properties>
+
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>org.lwjgl</groupId>
+                <artifactId>lwjgl-bom</artifactId>
+                <version>${lwjgl.version}</version>
+                <scope>import</scope>
+                <type>pom</type>
+            </dependency>
+        </dependencies>
+    </dependencyManagement>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.8.0</version>
+                <configuration>
+                    <source>1.8</source>
+                    <target>1.8</target>
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-jar-plugin</artifactId>
+                <version>3.2.0</version>
+                <configuration>
+                    <archive>
+                        <addMavenDescriptor>false</addMavenDescriptor>
+                        <manifest>
+                            <addClasspath>false</addClasspath>
+                        </manifest>
+                        <manifestEntries>
+                            <Main-Class>net.minecraft.launcher.Main</Main-Class>
+                            <Implementation-Version>${project.version}</Implementation-Version>
+                            <Multi-Release>true</Multi-Release> <!-- workaround to allow the launcher to run on newer java -->
+                        </manifestEntries>
+                    </archive>
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-shade-plugin</artifactId>
+                <version>3.2.4</version>
+                <executions>
+                    <execution>
+                        <phase>package</phase>
+                        <goals>
+                            <goal>shade</goal>
+                        </goals>
+                    </execution>
+                </executions>
+                <configuration>
+                    <minimizeJar>true</minimizeJar>
+                    <filters>
+                        <filter>
+                            <artifact>*log4j*:*log4j*</artifact>
+                            <includes>
+                                <include>**</include>
+                            </includes>
+                        </filter>
+                    </filters>
+                    <createDependencyReducedPom>false</createDependencyReducedPom>
+                    <transformers>
+                        <transformer
+                                implementation="io.github.edwgiz.log4j.maven.plugins.shade.transformer.Log4j2PluginCacheFileTransformer">
+                        </transformer>
+                    </transformers>
+                </configuration>
+                <dependencies>
+                    <dependency>
+                        <groupId>io.github.edwgiz</groupId>
+                        <artifactId>log4j-maven-shade-plugin-extensions</artifactId>
+                        <version>2.17.1</version>
+                    </dependency>
+                </dependencies>
+            </plugin>
+            <plugin>
+                <groupId>io.github.git-commit-id</groupId>
+                <artifactId>git-commit-id-maven-plugin</artifactId>
+                <version>5.0.0</version>
+                <executions>
+                    <execution>
+                        <id>inner-repo</id>
+                        <phase>initialize</phase>
+                        <goals>
+                            <goal>revision</goal>
+                        </goals>
+                        <configuration>
+                            <generateGitPropertiesFilename>${project.build.outputDirectory}/ver/git-inner.properties</generateGitPropertiesFilename>
+                            <includeOnlyProperties>
+                                <property>^git.dirty$</property>
+                            </includeOnlyProperties>
+                        </configuration>
+                    </execution>
+                    <execution>
+                        <id>outer-repo</id>
+                        <phase>initialize</phase>
+                        <goals>
+                            <goal>revision</goal>
+                        </goals>
+                        <configuration>
+                            <generateGitPropertiesFilename>${project.build.outputDirectory}/ver/git-outer.properties</generateGitPropertiesFilename>
+                            <dotGitDirectory>../.git</dotGitDirectory>
+                            <includeOnlyProperties>
+                                <property>^git.closest.tag.name$</property>
+                                <property>^git.closest.tag.commit.count$</property>
+                                <property>^git.dirty$</property>
+                                <property>^git.commit.id.abbrev$</property>
+                            </includeOnlyProperties>
+                        </configuration>
+                    </execution>
+                </executions>
+                <configuration>
+                    <generateGitPropertiesFile>true</generateGitPropertiesFile>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+
+    <repositories>
+        <repository>
+            <id>central</id>
+            <url>https://repo1.maven.org/maven2/</url>
+        </repository>
+        <repository>
+            <id>oss-snapshot</id>
+            <url>https://oss.sonatype.org/content/repositories/snapshots</url>
+            <snapshots>
+                <enabled>true</enabled>
+            </snapshots>
+        </repository>
+    </repositories>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-core</artifactId>
+            <version>2.17.2</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>commons-io</groupId>
+            <artifactId>commons-io</artifactId>
+            <version>2.11.0</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>commons-codec</groupId>
+            <artifactId>commons-codec</artifactId>
+            <version>1.15</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+            <version>3.12.0</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-compress</artifactId>
+            <version>1.21</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-text</artifactId>
+            <version>1.9</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.tukaani</groupId>
+            <artifactId>xz</artifactId>
+            <version>1.9</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.google.code.gson</groupId>
+            <artifactId>gson</artifactId>
+            <version>2.9.0</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+            <version>31.1-jre</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>net.sf.jopt-simple</groupId>
+            <artifactId>jopt-simple</artifactId>
+            <version>5.0.4</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.openjfx</groupId>
+            <artifactId>javafx-controls</artifactId>
+            <version>17</version>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.openjfx</groupId>
+            <artifactId>javafx-swing</artifactId>
+            <version>17</version>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.openjfx</groupId>
+            <artifactId>javafx-web</artifactId>
+            <version>17</version>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <version>1.18.22</version>
+            <scope>provided</scope>
+        </dependency>
+
+        <!--skin editor stuff-->
+        <!--<dependency>
+            <groupId>org.jogamp.jogl</groupId>
+            <artifactId>jogl-all-main</artifactId>
+            <version>2.3.2</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.jogamp.gluegen</groupId>
+            <artifactId>gluegen-rt-main</artifactId>
+            <version>2.3.2</version>
+            <scope>compile</scope>
+        </dependency>
+        -->
+        <dependency>
+            <groupId>org.joml</groupId>
+            <artifactId>joml</artifactId>
+            <version>1.10.4</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.lwjglx</groupId>
+            <artifactId>lwjgl3-awt</artifactId>
+            <version>0.1.9-SNAPSHOT</version>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.lwjgl</groupId>
+            <artifactId>lwjgl</artifactId>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.lwjgl</groupId>
+            <artifactId>lwjgl-opengl</artifactId>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.lwjgl</groupId>
+            <artifactId>lwjgl-stb</artifactId>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.lwjgl</groupId>
+            <artifactId>lwjgl</artifactId>
+            <classifier>${lwjgl.natives}</classifier>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.lwjgl</groupId>
+            <artifactId>lwjgl-opengl</artifactId>
+            <classifier>${lwjgl.natives}</classifier>
+            <scope>compile</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.lwjgl</groupId>
+            <artifactId>lwjgl-stb</artifactId>
+            <classifier>${lwjgl.natives}</classifier>
+            <scope>compile</scope>
+        </dependency>
+    </dependencies>
+
+    <profiles>
+        <profile>
+            <id>lwjgl-natives-linux-amd64</id>
+            <activation>
+                <os>
+                    <family>unix</family>
+                    <arch>amd64</arch>
+                </os>
+            </activation>
+            <properties>
+                <lwjgl.natives>natives-linux</lwjgl.natives>
+            </properties>
+        </profile>
+        <profile>
+            <id>lwjgl-natives-linux-aarch64</id>
+            <activation>
+                <os>
+                    <family>unix</family>
+                    <arch>aarch64</arch>
+                </os>
+            </activation>
+            <properties>
+                <lwjgl.natives>natives-linux-arm64</lwjgl.natives>
+            </properties>
+        </profile>
+        <profile>
+            <id>lwjgl-natives-linux-arm</id>
+            <activation>
+                <os>
+                    <family>unix</family>
+                    <arch>arm</arch>
+                </os>
+            </activation>
+            <properties>
+                <lwjgl.natives>natives-linux-arm32</lwjgl.natives>
+            </properties>
+        </profile>
+        <profile>
+            <id>lwjgl-natives-linux-arm32</id>
+            <activation>
+                <os>
+                    <family>unix</family>
+                    <arch>arm32</arch>
+                </os>
+            </activation>
+            <properties>
+                <lwjgl.natives>natives-linux-arm32</lwjgl.natives>
+            </properties>
+        </profile>
+        <profile>
+            <id>lwjgl-natives-macos-x86_64</id>
+            <activation>
+                <os>
+                    <family>mac</family>
+                    <arch>x86_64</arch>
+                </os>
+            </activation>
+            <properties>
+                <lwjgl.natives>natives-macos</lwjgl.natives>
+            </properties>
+            <dependencies>
+                <dependency>
+                    <groupId>org.lwjgl</groupId>
+                    <artifactId>lwjgl-vulkan</artifactId>
+                    <classifier>natives-macos</classifier>
+                </dependency>
+            </dependencies>
+        </profile>
+        <profile>
+            <id>lwjgl-natives-macos-aarch64</id>
+            <activation>
+                <os>
+                    <family>mac</family>
+                    <arch>aarch64</arch>
+                </os>
+            </activation>
+            <properties>
+                <lwjgl.natives>natives-macos-arm64</lwjgl.natives>
+            </properties>
+            <dependencies>
+                <dependency>
+                    <groupId>org.lwjgl</groupId>
+                    <artifactId>lwjgl-vulkan</artifactId>
+                    <classifier>natives-macos-arm64</classifier>
+                </dependency>
+            </dependencies>
+        </profile>
+        <profile>
+            <id>lwjgl-natives-windows-amd64</id>
+            <activation>
+                <os>
+                    <family>windows</family>
+                    <arch>amd64</arch>
+                </os>
+            </activation>
+            <properties>
+                <lwjgl.natives>natives-windows</lwjgl.natives>
+            </properties>
+        </profile>
+        <profile>
+            <id>lwjgl-natives-windows-x86</id>
+            <activation>
+                <os>
+                    <family>windows</family>
+                    <arch>x86</arch>
+                </os>
+            </activation>
+            <properties>
+                <lwjgl.natives>natives-windows-x86</lwjgl.natives>
+            </properties>
+        </profile>
+        <profile>
+            <id>lwjgl-natives-windows-aarch64</id>
+            <activation>
+                <os>
+                    <family>windows</family>
+                    <arch>aarch64</arch>
+                </os>
+            </activation>
+            <properties>
+                <lwjgl.natives>natives-windows-arm64</lwjgl.natives>
+            </properties>
+        </profile>
+    </profiles>
+</project>
diff --git a/com/mojang/authlib/Agent.java b/src/main/java/com/mojang/authlib/Agent.java
similarity index 100%
rename from com/mojang/authlib/Agent.java
rename to src/main/java/com/mojang/authlib/Agent.java
diff --git a/com/mojang/authlib/AuthenticationService.java b/src/main/java/com/mojang/authlib/AuthenticationService.java
similarity index 100%
rename from com/mojang/authlib/AuthenticationService.java
rename to src/main/java/com/mojang/authlib/AuthenticationService.java
diff --git a/com/mojang/authlib/BaseAuthenticationService.java b/src/main/java/com/mojang/authlib/BaseAuthenticationService.java
similarity index 100%
rename from com/mojang/authlib/BaseAuthenticationService.java
rename to src/main/java/com/mojang/authlib/BaseAuthenticationService.java
diff --git a/com/mojang/authlib/BaseUserAuthentication.java b/src/main/java/com/mojang/authlib/BaseUserAuthentication.java
similarity index 100%
rename from com/mojang/authlib/BaseUserAuthentication.java
rename to src/main/java/com/mojang/authlib/BaseUserAuthentication.java
diff --git a/com/mojang/authlib/GameProfile.java b/src/main/java/com/mojang/authlib/GameProfile.java
similarity index 100%
rename from com/mojang/authlib/GameProfile.java
rename to src/main/java/com/mojang/authlib/GameProfile.java
diff --git a/com/mojang/authlib/GameProfileRepository.java b/src/main/java/com/mojang/authlib/GameProfileRepository.java
similarity index 100%
rename from com/mojang/authlib/GameProfileRepository.java
rename to src/main/java/com/mojang/authlib/GameProfileRepository.java
diff --git a/com/mojang/authlib/HttpAuthenticationService.java b/src/main/java/com/mojang/authlib/HttpAuthenticationService.java
similarity index 100%
rename from com/mojang/authlib/HttpAuthenticationService.java
rename to src/main/java/com/mojang/authlib/HttpAuthenticationService.java
diff --git a/com/mojang/authlib/HttpUserAuthentication.java b/src/main/java/com/mojang/authlib/HttpUserAuthentication.java
similarity index 100%
rename from com/mojang/authlib/HttpUserAuthentication.java
rename to src/main/java/com/mojang/authlib/HttpUserAuthentication.java
diff --git a/com/mojang/authlib/ProfileLookupCallback.java b/src/main/java/com/mojang/authlib/ProfileLookupCallback.java
similarity index 100%
rename from com/mojang/authlib/ProfileLookupCallback.java
rename to src/main/java/com/mojang/authlib/ProfileLookupCallback.java
diff --git a/com/mojang/authlib/UserAuthentication.java b/src/main/java/com/mojang/authlib/UserAuthentication.java
similarity index 100%
rename from com/mojang/authlib/UserAuthentication.java
rename to src/main/java/com/mojang/authlib/UserAuthentication.java
diff --git a/com/mojang/authlib/UserType.java b/src/main/java/com/mojang/authlib/UserType.java
similarity index 100%
rename from com/mojang/authlib/UserType.java
rename to src/main/java/com/mojang/authlib/UserType.java
diff --git a/com/mojang/authlib/exceptions/AuthenticationException.java b/src/main/java/com/mojang/authlib/exceptions/AuthenticationException.java
similarity index 100%
rename from com/mojang/authlib/exceptions/AuthenticationException.java
rename to src/main/java/com/mojang/authlib/exceptions/AuthenticationException.java
diff --git a/com/mojang/authlib/exceptions/AuthenticationUnavailableException.java b/src/main/java/com/mojang/authlib/exceptions/AuthenticationUnavailableException.java
similarity index 100%
rename from com/mojang/authlib/exceptions/AuthenticationUnavailableException.java
rename to src/main/java/com/mojang/authlib/exceptions/AuthenticationUnavailableException.java
diff --git a/com/mojang/authlib/exceptions/InvalidCredentialsException.java b/src/main/java/com/mojang/authlib/exceptions/InvalidCredentialsException.java
similarity index 100%
rename from com/mojang/authlib/exceptions/InvalidCredentialsException.java
rename to src/main/java/com/mojang/authlib/exceptions/InvalidCredentialsException.java
diff --git a/com/mojang/authlib/exceptions/UserMigratedException.java b/src/main/java/com/mojang/authlib/exceptions/UserMigratedException.java
similarity index 100%
rename from com/mojang/authlib/exceptions/UserMigratedException.java
rename to src/main/java/com/mojang/authlib/exceptions/UserMigratedException.java
diff --git a/com/mojang/authlib/legacy/LegacyAuthenticationService.java b/src/main/java/com/mojang/authlib/legacy/LegacyAuthenticationService.java
similarity index 100%
rename from com/mojang/authlib/legacy/LegacyAuthenticationService.java
rename to src/main/java/com/mojang/authlib/legacy/LegacyAuthenticationService.java
diff --git a/com/mojang/authlib/legacy/LegacyMinecraftSessionService.java b/src/main/java/com/mojang/authlib/legacy/LegacyMinecraftSessionService.java
similarity index 100%
rename from com/mojang/authlib/legacy/LegacyMinecraftSessionService.java
rename to src/main/java/com/mojang/authlib/legacy/LegacyMinecraftSessionService.java
diff --git a/com/mojang/authlib/legacy/LegacyUserAuthentication.java b/src/main/java/com/mojang/authlib/legacy/LegacyUserAuthentication.java
similarity index 100%
rename from com/mojang/authlib/legacy/LegacyUserAuthentication.java
rename to src/main/java/com/mojang/authlib/legacy/LegacyUserAuthentication.java
diff --git a/com/mojang/authlib/minecraft/BaseMinecraftSessionService.java b/src/main/java/com/mojang/authlib/minecraft/BaseMinecraftSessionService.java
similarity index 100%
rename from com/mojang/authlib/minecraft/BaseMinecraftSessionService.java
rename to src/main/java/com/mojang/authlib/minecraft/BaseMinecraftSessionService.java
diff --git a/com/mojang/authlib/minecraft/HttpMinecraftSessionService.java b/src/main/java/com/mojang/authlib/minecraft/HttpMinecraftSessionService.java
similarity index 100%
rename from com/mojang/authlib/minecraft/HttpMinecraftSessionService.java
rename to src/main/java/com/mojang/authlib/minecraft/HttpMinecraftSessionService.java
diff --git a/com/mojang/authlib/minecraft/InsecureTextureException.java b/src/main/java/com/mojang/authlib/minecraft/InsecureTextureException.java
similarity index 100%
rename from com/mojang/authlib/minecraft/InsecureTextureException.java
rename to src/main/java/com/mojang/authlib/minecraft/InsecureTextureException.java
diff --git a/com/mojang/authlib/minecraft/MinecraftProfileTexture.java b/src/main/java/com/mojang/authlib/minecraft/MinecraftProfileTexture.java
similarity index 100%
rename from com/mojang/authlib/minecraft/MinecraftProfileTexture.java
rename to src/main/java/com/mojang/authlib/minecraft/MinecraftProfileTexture.java
diff --git a/com/mojang/authlib/minecraft/MinecraftSessionService.java b/src/main/java/com/mojang/authlib/minecraft/MinecraftSessionService.java
similarity index 100%
rename from com/mojang/authlib/minecraft/MinecraftSessionService.java
rename to src/main/java/com/mojang/authlib/minecraft/MinecraftSessionService.java
diff --git a/com/mojang/authlib/properties/Property.java b/src/main/java/com/mojang/authlib/properties/Property.java
similarity index 100%
rename from com/mojang/authlib/properties/Property.java
rename to src/main/java/com/mojang/authlib/properties/Property.java
diff --git a/com/mojang/authlib/properties/PropertyMap.java b/src/main/java/com/mojang/authlib/properties/PropertyMap.java
similarity index 100%
rename from com/mojang/authlib/properties/PropertyMap.java
rename to src/main/java/com/mojang/authlib/properties/PropertyMap.java
diff --git a/com/mojang/authlib/yggdrasil/ProfileIncompleteException.java b/src/main/java/com/mojang/authlib/yggdrasil/ProfileIncompleteException.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/ProfileIncompleteException.java
rename to src/main/java/com/mojang/authlib/yggdrasil/ProfileIncompleteException.java
diff --git a/com/mojang/authlib/yggdrasil/ProfileNotFoundException.java b/src/main/java/com/mojang/authlib/yggdrasil/ProfileNotFoundException.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/ProfileNotFoundException.java
rename to src/main/java/com/mojang/authlib/yggdrasil/ProfileNotFoundException.java
diff --git a/com/mojang/authlib/yggdrasil/YggdrasilAuthenticationService.java b/src/main/java/com/mojang/authlib/yggdrasil/YggdrasilAuthenticationService.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/YggdrasilAuthenticationService.java
rename to src/main/java/com/mojang/authlib/yggdrasil/YggdrasilAuthenticationService.java
diff --git a/com/mojang/authlib/yggdrasil/YggdrasilGameProfileRepository.java b/src/main/java/com/mojang/authlib/yggdrasil/YggdrasilGameProfileRepository.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/YggdrasilGameProfileRepository.java
rename to src/main/java/com/mojang/authlib/yggdrasil/YggdrasilGameProfileRepository.java
diff --git a/com/mojang/authlib/yggdrasil/YggdrasilMinecraftSessionService.java b/src/main/java/com/mojang/authlib/yggdrasil/YggdrasilMinecraftSessionService.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/YggdrasilMinecraftSessionService.java
rename to src/main/java/com/mojang/authlib/yggdrasil/YggdrasilMinecraftSessionService.java
diff --git a/com/mojang/authlib/yggdrasil/YggdrasilUserAuthentication.java b/src/main/java/com/mojang/authlib/yggdrasil/YggdrasilUserAuthentication.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/YggdrasilUserAuthentication.java
rename to src/main/java/com/mojang/authlib/yggdrasil/YggdrasilUserAuthentication.java
diff --git a/com/mojang/authlib/yggdrasil/request/AuthenticationRequest.java b/src/main/java/com/mojang/authlib/yggdrasil/request/AuthenticationRequest.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/request/AuthenticationRequest.java
rename to src/main/java/com/mojang/authlib/yggdrasil/request/AuthenticationRequest.java
diff --git a/com/mojang/authlib/yggdrasil/request/InvalidateRequest.java b/src/main/java/com/mojang/authlib/yggdrasil/request/InvalidateRequest.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/request/InvalidateRequest.java
rename to src/main/java/com/mojang/authlib/yggdrasil/request/InvalidateRequest.java
diff --git a/com/mojang/authlib/yggdrasil/request/JoinMinecraftServerRequest.java b/src/main/java/com/mojang/authlib/yggdrasil/request/JoinMinecraftServerRequest.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/request/JoinMinecraftServerRequest.java
rename to src/main/java/com/mojang/authlib/yggdrasil/request/JoinMinecraftServerRequest.java
diff --git a/com/mojang/authlib/yggdrasil/request/RefreshRequest.java b/src/main/java/com/mojang/authlib/yggdrasil/request/RefreshRequest.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/request/RefreshRequest.java
rename to src/main/java/com/mojang/authlib/yggdrasil/request/RefreshRequest.java
diff --git a/com/mojang/authlib/yggdrasil/request/ValidateRequest.java b/src/main/java/com/mojang/authlib/yggdrasil/request/ValidateRequest.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/request/ValidateRequest.java
rename to src/main/java/com/mojang/authlib/yggdrasil/request/ValidateRequest.java
diff --git a/com/mojang/authlib/yggdrasil/response/AuthenticationResponse.java b/src/main/java/com/mojang/authlib/yggdrasil/response/AuthenticationResponse.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/response/AuthenticationResponse.java
rename to src/main/java/com/mojang/authlib/yggdrasil/response/AuthenticationResponse.java
diff --git a/com/mojang/authlib/yggdrasil/response/HasJoinedMinecraftServerResponse.java b/src/main/java/com/mojang/authlib/yggdrasil/response/HasJoinedMinecraftServerResponse.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/response/HasJoinedMinecraftServerResponse.java
rename to src/main/java/com/mojang/authlib/yggdrasil/response/HasJoinedMinecraftServerResponse.java
diff --git a/com/mojang/authlib/yggdrasil/response/MinecraftProfilePropertiesResponse.java b/src/main/java/com/mojang/authlib/yggdrasil/response/MinecraftProfilePropertiesResponse.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/response/MinecraftProfilePropertiesResponse.java
rename to src/main/java/com/mojang/authlib/yggdrasil/response/MinecraftProfilePropertiesResponse.java
diff --git a/com/mojang/authlib/yggdrasil/response/MinecraftTexturesPayload.java b/src/main/java/com/mojang/authlib/yggdrasil/response/MinecraftTexturesPayload.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/response/MinecraftTexturesPayload.java
rename to src/main/java/com/mojang/authlib/yggdrasil/response/MinecraftTexturesPayload.java
diff --git a/com/mojang/authlib/yggdrasil/response/ProfileSearchResultsResponse.java b/src/main/java/com/mojang/authlib/yggdrasil/response/ProfileSearchResultsResponse.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/response/ProfileSearchResultsResponse.java
rename to src/main/java/com/mojang/authlib/yggdrasil/response/ProfileSearchResultsResponse.java
diff --git a/com/mojang/authlib/yggdrasil/response/RefreshResponse.java b/src/main/java/com/mojang/authlib/yggdrasil/response/RefreshResponse.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/response/RefreshResponse.java
rename to src/main/java/com/mojang/authlib/yggdrasil/response/RefreshResponse.java
diff --git a/com/mojang/authlib/yggdrasil/response/Response.java b/src/main/java/com/mojang/authlib/yggdrasil/response/Response.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/response/Response.java
rename to src/main/java/com/mojang/authlib/yggdrasil/response/Response.java
diff --git a/com/mojang/authlib/yggdrasil/response/User.java b/src/main/java/com/mojang/authlib/yggdrasil/response/User.java
similarity index 100%
rename from com/mojang/authlib/yggdrasil/response/User.java
rename to src/main/java/com/mojang/authlib/yggdrasil/response/User.java
diff --git a/com/mojang/launcher/Http.java b/src/main/java/com/mojang/launcher/Http.java
similarity index 100%
rename from com/mojang/launcher/Http.java
rename to src/main/java/com/mojang/launcher/Http.java
diff --git a/com/mojang/launcher/Launcher.java b/src/main/java/com/mojang/launcher/Launcher.java
similarity index 100%
rename from com/mojang/launcher/Launcher.java
rename to src/main/java/com/mojang/launcher/Launcher.java
diff --git a/com/mojang/launcher/LegacyPropertyMapSerializer.java b/src/main/java/com/mojang/launcher/LegacyPropertyMapSerializer.java
similarity index 100%
rename from com/mojang/launcher/LegacyPropertyMapSerializer.java
rename to src/main/java/com/mojang/launcher/LegacyPropertyMapSerializer.java
diff --git a/com/mojang/launcher/OperatingSystem.java b/src/main/java/com/mojang/launcher/OperatingSystem.java
similarity index 100%
rename from com/mojang/launcher/OperatingSystem.java
rename to src/main/java/com/mojang/launcher/OperatingSystem.java
diff --git a/com/mojang/launcher/SharedLauncherConstants.java b/src/main/java/com/mojang/launcher/SharedLauncherConstants.java
similarity index 100%
rename from com/mojang/launcher/SharedLauncherConstants.java
rename to src/main/java/com/mojang/launcher/SharedLauncherConstants.java
diff --git a/com/mojang/launcher/UserInterface.java b/src/main/java/com/mojang/launcher/UserInterface.java
similarity index 100%
rename from com/mojang/launcher/UserInterface.java
rename to src/main/java/com/mojang/launcher/UserInterface.java
diff --git a/com/mojang/launcher/events/GameOutputLogProcessor.java b/src/main/java/com/mojang/launcher/events/GameOutputLogProcessor.java
similarity index 100%
rename from com/mojang/launcher/events/GameOutputLogProcessor.java
rename to src/main/java/com/mojang/launcher/events/GameOutputLogProcessor.java
diff --git a/com/mojang/launcher/events/RefreshedVersionsListener.java b/src/main/java/com/mojang/launcher/events/RefreshedVersionsListener.java
similarity index 100%
rename from com/mojang/launcher/events/RefreshedVersionsListener.java
rename to src/main/java/com/mojang/launcher/events/RefreshedVersionsListener.java
diff --git a/com/mojang/launcher/game/GameInstanceStatus.java b/src/main/java/com/mojang/launcher/game/GameInstanceStatus.java
similarity index 100%
rename from com/mojang/launcher/game/GameInstanceStatus.java
rename to src/main/java/com/mojang/launcher/game/GameInstanceStatus.java
diff --git a/com/mojang/launcher/game/process/AbstractGameProcess.java b/src/main/java/com/mojang/launcher/game/process/AbstractGameProcess.java
similarity index 100%
rename from com/mojang/launcher/game/process/AbstractGameProcess.java
rename to src/main/java/com/mojang/launcher/game/process/AbstractGameProcess.java
diff --git a/com/mojang/launcher/game/process/GameProcess.java b/src/main/java/com/mojang/launcher/game/process/GameProcess.java
similarity index 100%
rename from com/mojang/launcher/game/process/GameProcess.java
rename to src/main/java/com/mojang/launcher/game/process/GameProcess.java
diff --git a/com/mojang/launcher/game/process/GameProcessBuilder.java b/src/main/java/com/mojang/launcher/game/process/GameProcessBuilder.java
similarity index 100%
rename from com/mojang/launcher/game/process/GameProcessBuilder.java
rename to src/main/java/com/mojang/launcher/game/process/GameProcessBuilder.java
diff --git a/com/mojang/launcher/game/process/GameProcessFactory.java b/src/main/java/com/mojang/launcher/game/process/GameProcessFactory.java
similarity index 100%
rename from com/mojang/launcher/game/process/GameProcessFactory.java
rename to src/main/java/com/mojang/launcher/game/process/GameProcessFactory.java
diff --git a/com/mojang/launcher/game/process/GameProcessRunnable.java b/src/main/java/com/mojang/launcher/game/process/GameProcessRunnable.java
similarity index 100%
rename from com/mojang/launcher/game/process/GameProcessRunnable.java
rename to src/main/java/com/mojang/launcher/game/process/GameProcessRunnable.java
diff --git a/com/mojang/launcher/game/process/direct/DirectGameProcess.java b/src/main/java/com/mojang/launcher/game/process/direct/DirectGameProcess.java
similarity index 100%
rename from com/mojang/launcher/game/process/direct/DirectGameProcess.java
rename to src/main/java/com/mojang/launcher/game/process/direct/DirectGameProcess.java
diff --git a/com/mojang/launcher/game/process/direct/DirectGameProcessFactory.java b/src/main/java/com/mojang/launcher/game/process/direct/DirectGameProcessFactory.java
similarity index 100%
rename from com/mojang/launcher/game/process/direct/DirectGameProcessFactory.java
rename to src/main/java/com/mojang/launcher/game/process/direct/DirectGameProcessFactory.java
diff --git a/com/mojang/launcher/game/process/direct/DirectProcessInputMonitor.java b/src/main/java/com/mojang/launcher/game/process/direct/DirectProcessInputMonitor.java
similarity index 100%
rename from com/mojang/launcher/game/process/direct/DirectProcessInputMonitor.java
rename to src/main/java/com/mojang/launcher/game/process/direct/DirectProcessInputMonitor.java
diff --git a/com/mojang/launcher/game/runner/AbstractGameRunner.java b/src/main/java/com/mojang/launcher/game/runner/AbstractGameRunner.java
similarity index 100%
rename from com/mojang/launcher/game/runner/AbstractGameRunner.java
rename to src/main/java/com/mojang/launcher/game/runner/AbstractGameRunner.java
diff --git a/com/mojang/launcher/game/runner/GameRunner.java b/src/main/java/com/mojang/launcher/game/runner/GameRunner.java
similarity index 100%
rename from com/mojang/launcher/game/runner/GameRunner.java
rename to src/main/java/com/mojang/launcher/game/runner/GameRunner.java
diff --git a/com/mojang/launcher/game/runner/GameRunnerListener.java b/src/main/java/com/mojang/launcher/game/runner/GameRunnerListener.java
similarity index 100%
rename from com/mojang/launcher/game/runner/GameRunnerListener.java
rename to src/main/java/com/mojang/launcher/game/runner/GameRunnerListener.java
diff --git a/com/mojang/launcher/updater/DateTypeAdapter.java b/src/main/java/com/mojang/launcher/updater/DateTypeAdapter.java
similarity index 100%
rename from com/mojang/launcher/updater/DateTypeAdapter.java
rename to src/main/java/com/mojang/launcher/updater/DateTypeAdapter.java
diff --git a/com/mojang/launcher/updater/DownloadProgress.java b/src/main/java/com/mojang/launcher/updater/DownloadProgress.java
similarity index 100%
rename from com/mojang/launcher/updater/DownloadProgress.java
rename to src/main/java/com/mojang/launcher/updater/DownloadProgress.java
diff --git a/com/mojang/launcher/updater/ExceptionalThreadPoolExecutor.java b/src/main/java/com/mojang/launcher/updater/ExceptionalThreadPoolExecutor.java
similarity index 100%
rename from com/mojang/launcher/updater/ExceptionalThreadPoolExecutor.java
rename to src/main/java/com/mojang/launcher/updater/ExceptionalThreadPoolExecutor.java
diff --git a/com/mojang/launcher/updater/FileTypeAdapter.java b/src/main/java/com/mojang/launcher/updater/FileTypeAdapter.java
similarity index 100%
rename from com/mojang/launcher/updater/FileTypeAdapter.java
rename to src/main/java/com/mojang/launcher/updater/FileTypeAdapter.java
diff --git a/com/mojang/launcher/updater/LowerCaseEnumTypeAdapterFactory.java b/src/main/java/com/mojang/launcher/updater/LowerCaseEnumTypeAdapterFactory.java
similarity index 100%
rename from com/mojang/launcher/updater/LowerCaseEnumTypeAdapterFactory.java
rename to src/main/java/com/mojang/launcher/updater/LowerCaseEnumTypeAdapterFactory.java
diff --git a/com/mojang/launcher/updater/VersionFilter.java b/src/main/java/com/mojang/launcher/updater/VersionFilter.java
similarity index 100%
rename from com/mojang/launcher/updater/VersionFilter.java
rename to src/main/java/com/mojang/launcher/updater/VersionFilter.java
diff --git a/com/mojang/launcher/updater/VersionManager.java b/src/main/java/com/mojang/launcher/updater/VersionManager.java
similarity index 100%
rename from com/mojang/launcher/updater/VersionManager.java
rename to src/main/java/com/mojang/launcher/updater/VersionManager.java
diff --git a/com/mojang/launcher/updater/VersionSyncInfo.java b/src/main/java/com/mojang/launcher/updater/VersionSyncInfo.java
similarity index 100%
rename from com/mojang/launcher/updater/VersionSyncInfo.java
rename to src/main/java/com/mojang/launcher/updater/VersionSyncInfo.java
diff --git a/com/mojang/launcher/updater/download/ChecksummedDownloadable.java b/src/main/java/com/mojang/launcher/updater/download/ChecksummedDownloadable.java
similarity index 100%
rename from com/mojang/launcher/updater/download/ChecksummedDownloadable.java
rename to src/main/java/com/mojang/launcher/updater/download/ChecksummedDownloadable.java
diff --git a/com/mojang/launcher/updater/download/DownloadJob.java b/src/main/java/com/mojang/launcher/updater/download/DownloadJob.java
similarity index 100%
rename from com/mojang/launcher/updater/download/DownloadJob.java
rename to src/main/java/com/mojang/launcher/updater/download/DownloadJob.java
diff --git a/com/mojang/launcher/updater/download/DownloadListener.java b/src/main/java/com/mojang/launcher/updater/download/DownloadListener.java
similarity index 100%
rename from com/mojang/launcher/updater/download/DownloadListener.java
rename to src/main/java/com/mojang/launcher/updater/download/DownloadListener.java
diff --git a/com/mojang/launcher/updater/download/Downloadable.java b/src/main/java/com/mojang/launcher/updater/download/Downloadable.java
similarity index 100%
rename from com/mojang/launcher/updater/download/Downloadable.java
rename to src/main/java/com/mojang/launcher/updater/download/Downloadable.java
diff --git a/com/mojang/launcher/updater/download/EtagDownloadable.java b/src/main/java/com/mojang/launcher/updater/download/EtagDownloadable.java
similarity index 100%
rename from com/mojang/launcher/updater/download/EtagDownloadable.java
rename to src/main/java/com/mojang/launcher/updater/download/EtagDownloadable.java
diff --git a/com/mojang/launcher/updater/download/MonitoringInputStream.java b/src/main/java/com/mojang/launcher/updater/download/MonitoringInputStream.java
similarity index 100%
rename from com/mojang/launcher/updater/download/MonitoringInputStream.java
rename to src/main/java/com/mojang/launcher/updater/download/MonitoringInputStream.java
diff --git a/com/mojang/launcher/updater/download/ProgressContainer.java b/src/main/java/com/mojang/launcher/updater/download/ProgressContainer.java
similarity index 100%
rename from com/mojang/launcher/updater/download/ProgressContainer.java
rename to src/main/java/com/mojang/launcher/updater/download/ProgressContainer.java
diff --git a/com/mojang/launcher/updater/download/assets/AssetDownloadable.java b/src/main/java/com/mojang/launcher/updater/download/assets/AssetDownloadable.java
similarity index 100%
rename from com/mojang/launcher/updater/download/assets/AssetDownloadable.java
rename to src/main/java/com/mojang/launcher/updater/download/assets/AssetDownloadable.java
diff --git a/com/mojang/launcher/updater/download/assets/AssetIndex.java b/src/main/java/com/mojang/launcher/updater/download/assets/AssetIndex.java
similarity index 100%
rename from com/mojang/launcher/updater/download/assets/AssetIndex.java
rename to src/main/java/com/mojang/launcher/updater/download/assets/AssetIndex.java
diff --git a/com/mojang/launcher/versions/CompatibilityRule.java b/src/main/java/com/mojang/launcher/versions/CompatibilityRule.java
similarity index 100%
rename from com/mojang/launcher/versions/CompatibilityRule.java
rename to src/main/java/com/mojang/launcher/versions/CompatibilityRule.java
diff --git a/com/mojang/launcher/versions/CompleteVersion.java b/src/main/java/com/mojang/launcher/versions/CompleteVersion.java
similarity index 100%
rename from com/mojang/launcher/versions/CompleteVersion.java
rename to src/main/java/com/mojang/launcher/versions/CompleteVersion.java
diff --git a/com/mojang/launcher/versions/ExtractRules.java b/src/main/java/com/mojang/launcher/versions/ExtractRules.java
similarity index 100%
rename from com/mojang/launcher/versions/ExtractRules.java
rename to src/main/java/com/mojang/launcher/versions/ExtractRules.java
diff --git a/com/mojang/launcher/versions/ReleaseType.java b/src/main/java/com/mojang/launcher/versions/ReleaseType.java
similarity index 100%
rename from com/mojang/launcher/versions/ReleaseType.java
rename to src/main/java/com/mojang/launcher/versions/ReleaseType.java
diff --git a/com/mojang/launcher/versions/ReleaseTypeAdapterFactory.java b/src/main/java/com/mojang/launcher/versions/ReleaseTypeAdapterFactory.java
similarity index 100%
rename from com/mojang/launcher/versions/ReleaseTypeAdapterFactory.java
rename to src/main/java/com/mojang/launcher/versions/ReleaseTypeAdapterFactory.java
diff --git a/com/mojang/launcher/versions/ReleaseTypeFactory.java b/src/main/java/com/mojang/launcher/versions/ReleaseTypeFactory.java
similarity index 100%
rename from com/mojang/launcher/versions/ReleaseTypeFactory.java
rename to src/main/java/com/mojang/launcher/versions/ReleaseTypeFactory.java
diff --git a/com/mojang/launcher/versions/Version.java b/src/main/java/com/mojang/launcher/versions/Version.java
similarity index 100%
rename from com/mojang/launcher/versions/Version.java
rename to src/main/java/com/mojang/launcher/versions/Version.java
diff --git a/com/mojang/util/QueueLogAppender.java b/src/main/java/com/mojang/util/QueueLogAppender.java
similarity index 100%
rename from com/mojang/util/QueueLogAppender.java
rename to src/main/java/com/mojang/util/QueueLogAppender.java
diff --git a/com/mojang/util/UUIDTypeAdapter.java b/src/main/java/com/mojang/util/UUIDTypeAdapter.java
similarity index 100%
rename from com/mojang/util/UUIDTypeAdapter.java
rename to src/main/java/com/mojang/util/UUIDTypeAdapter.java
diff --git a/net/minecraft/hopper/Crash.java b/src/main/java/net/minecraft/hopper/Crash.java
similarity index 100%
rename from net/minecraft/hopper/Crash.java
rename to src/main/java/net/minecraft/hopper/Crash.java
diff --git a/net/minecraft/hopper/HopperService.java b/src/main/java/net/minecraft/hopper/HopperService.java
similarity index 100%
rename from net/minecraft/hopper/HopperService.java
rename to src/main/java/net/minecraft/hopper/HopperService.java
diff --git a/net/minecraft/hopper/Problem.java b/src/main/java/net/minecraft/hopper/Problem.java
similarity index 100%
rename from net/minecraft/hopper/Problem.java
rename to src/main/java/net/minecraft/hopper/Problem.java
diff --git a/net/minecraft/hopper/PublishRequest.java b/src/main/java/net/minecraft/hopper/PublishRequest.java
similarity index 100%
rename from net/minecraft/hopper/PublishRequest.java
rename to src/main/java/net/minecraft/hopper/PublishRequest.java
diff --git a/net/minecraft/hopper/PublishResponse.java b/src/main/java/net/minecraft/hopper/PublishResponse.java
similarity index 100%
rename from net/minecraft/hopper/PublishResponse.java
rename to src/main/java/net/minecraft/hopper/PublishResponse.java
diff --git a/net/minecraft/hopper/Report.java b/src/main/java/net/minecraft/hopper/Report.java
similarity index 100%
rename from net/minecraft/hopper/Report.java
rename to src/main/java/net/minecraft/hopper/Report.java
diff --git a/net/minecraft/hopper/Response.java b/src/main/java/net/minecraft/hopper/Response.java
similarity index 100%
rename from net/minecraft/hopper/Response.java
rename to src/main/java/net/minecraft/hopper/Response.java
diff --git a/net/minecraft/hopper/SubmitRequest.java b/src/main/java/net/minecraft/hopper/SubmitRequest.java
similarity index 100%
rename from net/minecraft/hopper/SubmitRequest.java
rename to src/main/java/net/minecraft/hopper/SubmitRequest.java
diff --git a/net/minecraft/hopper/SubmitResponse.java b/src/main/java/net/minecraft/hopper/SubmitResponse.java
similarity index 100%
rename from net/minecraft/hopper/SubmitResponse.java
rename to src/main/java/net/minecraft/hopper/SubmitResponse.java
diff --git a/net/minecraft/hopper/Util.java b/src/main/java/net/minecraft/hopper/Util.java
similarity index 100%
rename from net/minecraft/hopper/Util.java
rename to src/main/java/net/minecraft/hopper/Util.java
diff --git a/net/minecraft/launcher/CompatibilityRule.java b/src/main/java/net/minecraft/launcher/CompatibilityRule.java
similarity index 100%
rename from net/minecraft/launcher/CompatibilityRule.java
rename to src/main/java/net/minecraft/launcher/CompatibilityRule.java
diff --git a/net/minecraft/launcher/CurrentLaunchFeatureMatcher.java b/src/main/java/net/minecraft/launcher/CurrentLaunchFeatureMatcher.java
similarity index 100%
rename from net/minecraft/launcher/CurrentLaunchFeatureMatcher.java
rename to src/main/java/net/minecraft/launcher/CurrentLaunchFeatureMatcher.java
diff --git a/net/minecraft/launcher/Launcher.java b/src/main/java/net/minecraft/launcher/Launcher.java
similarity index 100%
rename from net/minecraft/launcher/Launcher.java
rename to src/main/java/net/minecraft/launcher/Launcher.java
diff --git a/net/minecraft/launcher/LauncherConstants.java b/src/main/java/net/minecraft/launcher/LauncherConstants.java
similarity index 100%
rename from net/minecraft/launcher/LauncherConstants.java
rename to src/main/java/net/minecraft/launcher/LauncherConstants.java
diff --git a/net/minecraft/launcher/Main.java b/src/main/java/net/minecraft/launcher/Main.java
similarity index 100%
rename from net/minecraft/launcher/Main.java
rename to src/main/java/net/minecraft/launcher/Main.java
diff --git a/net/minecraft/launcher/MinecraftUserInterface.java b/src/main/java/net/minecraft/launcher/MinecraftUserInterface.java
similarity index 100%
rename from net/minecraft/launcher/MinecraftUserInterface.java
rename to src/main/java/net/minecraft/launcher/MinecraftUserInterface.java
diff --git a/net/minecraft/launcher/SwingUserInterface.java b/src/main/java/net/minecraft/launcher/SwingUserInterface.java
similarity index 100%
rename from net/minecraft/launcher/SwingUserInterface.java
rename to src/main/java/net/minecraft/launcher/SwingUserInterface.java
diff --git a/net/minecraft/launcher/game/GameLaunchDispatcher.java b/src/main/java/net/minecraft/launcher/game/GameLaunchDispatcher.java
similarity index 100%
rename from net/minecraft/launcher/game/GameLaunchDispatcher.java
rename to src/main/java/net/minecraft/launcher/game/GameLaunchDispatcher.java
diff --git a/net/minecraft/launcher/game/MinecraftGameRunner.java b/src/main/java/net/minecraft/launcher/game/MinecraftGameRunner.java
similarity index 100%
rename from net/minecraft/launcher/game/MinecraftGameRunner.java
rename to src/main/java/net/minecraft/launcher/game/MinecraftGameRunner.java
diff --git a/net/minecraft/launcher/game/MinecraftReleaseType.java b/src/main/java/net/minecraft/launcher/game/MinecraftReleaseType.java
similarity index 100%
rename from net/minecraft/launcher/game/MinecraftReleaseType.java
rename to src/main/java/net/minecraft/launcher/game/MinecraftReleaseType.java
diff --git a/net/minecraft/launcher/game/MinecraftReleaseTypeFactory.java b/src/main/java/net/minecraft/launcher/game/MinecraftReleaseTypeFactory.java
similarity index 100%
rename from net/minecraft/launcher/game/MinecraftReleaseTypeFactory.java
rename to src/main/java/net/minecraft/launcher/game/MinecraftReleaseTypeFactory.java
diff --git a/net/minecraft/launcher/profile/AuthenticationDatabase.java b/src/main/java/net/minecraft/launcher/profile/AuthenticationDatabase.java
similarity index 100%
rename from net/minecraft/launcher/profile/AuthenticationDatabase.java
rename to src/main/java/net/minecraft/launcher/profile/AuthenticationDatabase.java
diff --git a/net/minecraft/launcher/profile/LauncherVisibilityRule.java b/src/main/java/net/minecraft/launcher/profile/LauncherVisibilityRule.java
similarity index 100%
rename from net/minecraft/launcher/profile/LauncherVisibilityRule.java
rename to src/main/java/net/minecraft/launcher/profile/LauncherVisibilityRule.java
diff --git a/net/minecraft/launcher/profile/Profile.java b/src/main/java/net/minecraft/launcher/profile/Profile.java
similarity index 100%
rename from net/minecraft/launcher/profile/Profile.java
rename to src/main/java/net/minecraft/launcher/profile/Profile.java
diff --git a/net/minecraft/launcher/profile/ProfileManager.java b/src/main/java/net/minecraft/launcher/profile/ProfileManager.java
similarity index 100%
rename from net/minecraft/launcher/profile/ProfileManager.java
rename to src/main/java/net/minecraft/launcher/profile/ProfileManager.java
diff --git a/net/minecraft/launcher/profile/RefreshedProfilesListener.java b/src/main/java/net/minecraft/launcher/profile/RefreshedProfilesListener.java
similarity index 100%
rename from net/minecraft/launcher/profile/RefreshedProfilesListener.java
rename to src/main/java/net/minecraft/launcher/profile/RefreshedProfilesListener.java
diff --git a/net/minecraft/launcher/profile/UserChangedListener.java b/src/main/java/net/minecraft/launcher/profile/UserChangedListener.java
similarity index 100%
rename from net/minecraft/launcher/profile/UserChangedListener.java
rename to src/main/java/net/minecraft/launcher/profile/UserChangedListener.java
diff --git a/net/minecraft/launcher/ui/BottomBarPanel.java b/src/main/java/net/minecraft/launcher/ui/BottomBarPanel.java
similarity index 100%
rename from net/minecraft/launcher/ui/BottomBarPanel.java
rename to src/main/java/net/minecraft/launcher/ui/BottomBarPanel.java
diff --git a/net/minecraft/launcher/ui/LauncherPanel.java b/src/main/java/net/minecraft/launcher/ui/LauncherPanel.java
similarity index 100%
rename from net/minecraft/launcher/ui/LauncherPanel.java
rename to src/main/java/net/minecraft/launcher/ui/LauncherPanel.java
diff --git a/net/minecraft/launcher/ui/TexturedPanel.java b/src/main/java/net/minecraft/launcher/ui/TexturedPanel.java
similarity index 100%
rename from net/minecraft/launcher/ui/TexturedPanel.java
rename to src/main/java/net/minecraft/launcher/ui/TexturedPanel.java
diff --git a/net/minecraft/launcher/ui/bottombar/PlayButtonPanel.java b/src/main/java/net/minecraft/launcher/ui/bottombar/PlayButtonPanel.java
similarity index 100%
rename from net/minecraft/launcher/ui/bottombar/PlayButtonPanel.java
rename to src/main/java/net/minecraft/launcher/ui/bottombar/PlayButtonPanel.java
diff --git a/net/minecraft/launcher/ui/bottombar/PlayerInfoPanel.java b/src/main/java/net/minecraft/launcher/ui/bottombar/PlayerInfoPanel.java
similarity index 100%
rename from net/minecraft/launcher/ui/bottombar/PlayerInfoPanel.java
rename to src/main/java/net/minecraft/launcher/ui/bottombar/PlayerInfoPanel.java
diff --git a/net/minecraft/launcher/ui/bottombar/ProfileSelectionPanel.java b/src/main/java/net/minecraft/launcher/ui/bottombar/ProfileSelectionPanel.java
similarity index 100%
rename from net/minecraft/launcher/ui/bottombar/ProfileSelectionPanel.java
rename to src/main/java/net/minecraft/launcher/ui/bottombar/ProfileSelectionPanel.java
diff --git a/net/minecraft/launcher/ui/bottombar/SidebarGridForm.java b/src/main/java/net/minecraft/launcher/ui/bottombar/SidebarGridForm.java
similarity index 100%
rename from net/minecraft/launcher/ui/bottombar/SidebarGridForm.java
rename to src/main/java/net/minecraft/launcher/ui/bottombar/SidebarGridForm.java
diff --git a/net/minecraft/launcher/ui/bottombar/StatusPanelForm.java b/src/main/java/net/minecraft/launcher/ui/bottombar/StatusPanelForm.java
similarity index 100%
rename from net/minecraft/launcher/ui/bottombar/StatusPanelForm.java
rename to src/main/java/net/minecraft/launcher/ui/bottombar/StatusPanelForm.java
diff --git a/net/minecraft/launcher/ui/popups/login/AuthErrorForm.java b/src/main/java/net/minecraft/launcher/ui/popups/login/AuthErrorForm.java
similarity index 100%
rename from net/minecraft/launcher/ui/popups/login/AuthErrorForm.java
rename to src/main/java/net/minecraft/launcher/ui/popups/login/AuthErrorForm.java
diff --git a/net/minecraft/launcher/ui/popups/login/ExistingUserListForm.java b/src/main/java/net/minecraft/launcher/ui/popups/login/ExistingUserListForm.java
similarity index 100%
rename from net/minecraft/launcher/ui/popups/login/ExistingUserListForm.java
rename to src/main/java/net/minecraft/launcher/ui/popups/login/ExistingUserListForm.java
diff --git a/net/minecraft/launcher/ui/popups/login/LogInForm.java b/src/main/java/net/minecraft/launcher/ui/popups/login/LogInForm.java
similarity index 100%
rename from net/minecraft/launcher/ui/popups/login/LogInForm.java
rename to src/main/java/net/minecraft/launcher/ui/popups/login/LogInForm.java
diff --git a/net/minecraft/launcher/ui/popups/login/LogInPopup.java b/src/main/java/net/minecraft/launcher/ui/popups/login/LogInPopup.java
similarity index 100%
rename from net/minecraft/launcher/ui/popups/login/LogInPopup.java
rename to src/main/java/net/minecraft/launcher/ui/popups/login/LogInPopup.java
diff --git a/net/minecraft/launcher/ui/popups/profile/ProfileEditorPopup.java b/src/main/java/net/minecraft/launcher/ui/popups/profile/ProfileEditorPopup.java
similarity index 100%
rename from net/minecraft/launcher/ui/popups/profile/ProfileEditorPopup.java
rename to src/main/java/net/minecraft/launcher/ui/popups/profile/ProfileEditorPopup.java
diff --git a/net/minecraft/launcher/ui/popups/profile/ProfileInfoPanel.java b/src/main/java/net/minecraft/launcher/ui/popups/profile/ProfileInfoPanel.java
similarity index 100%
rename from net/minecraft/launcher/ui/popups/profile/ProfileInfoPanel.java
rename to src/main/java/net/minecraft/launcher/ui/popups/profile/ProfileInfoPanel.java
diff --git a/net/minecraft/launcher/ui/popups/profile/ProfileJavaPanel.java b/src/main/java/net/minecraft/launcher/ui/popups/profile/ProfileJavaPanel.java
similarity index 100%
rename from net/minecraft/launcher/ui/popups/profile/ProfileJavaPanel.java
rename to src/main/java/net/minecraft/launcher/ui/popups/profile/ProfileJavaPanel.java
diff --git a/net/minecraft/launcher/ui/popups/profile/ProfileVersionPanel.java b/src/main/java/net/minecraft/launcher/ui/popups/profile/ProfileVersionPanel.java
similarity index 100%
rename from net/minecraft/launcher/ui/popups/profile/ProfileVersionPanel.java
rename to src/main/java/net/minecraft/launcher/ui/popups/profile/ProfileVersionPanel.java
diff --git a/net/minecraft/launcher/ui/tabs/ConsoleTab.java b/src/main/java/net/minecraft/launcher/ui/tabs/ConsoleTab.java
similarity index 100%
rename from net/minecraft/launcher/ui/tabs/ConsoleTab.java
rename to src/main/java/net/minecraft/launcher/ui/tabs/ConsoleTab.java
diff --git a/net/minecraft/launcher/ui/tabs/CrashReportTab.java b/src/main/java/net/minecraft/launcher/ui/tabs/CrashReportTab.java
similarity index 100%
rename from net/minecraft/launcher/ui/tabs/CrashReportTab.java
rename to src/main/java/net/minecraft/launcher/ui/tabs/CrashReportTab.java
diff --git a/net/minecraft/launcher/ui/tabs/GameOutputTab.java b/src/main/java/net/minecraft/launcher/ui/tabs/GameOutputTab.java
similarity index 100%
rename from net/minecraft/launcher/ui/tabs/GameOutputTab.java
rename to src/main/java/net/minecraft/launcher/ui/tabs/GameOutputTab.java
diff --git a/net/minecraft/launcher/ui/tabs/LauncherTabPanel.java b/src/main/java/net/minecraft/launcher/ui/tabs/LauncherTabPanel.java
similarity index 100%
rename from net/minecraft/launcher/ui/tabs/LauncherTabPanel.java
rename to src/main/java/net/minecraft/launcher/ui/tabs/LauncherTabPanel.java
diff --git a/net/minecraft/launcher/ui/tabs/ProfileListTab.java b/src/main/java/net/minecraft/launcher/ui/tabs/ProfileListTab.java
similarity index 100%
rename from net/minecraft/launcher/ui/tabs/ProfileListTab.java
rename to src/main/java/net/minecraft/launcher/ui/tabs/ProfileListTab.java
diff --git a/net/minecraft/launcher/ui/tabs/WebsiteTab.java b/src/main/java/net/minecraft/launcher/ui/tabs/WebsiteTab.java
similarity index 100%
rename from net/minecraft/launcher/ui/tabs/WebsiteTab.java
rename to src/main/java/net/minecraft/launcher/ui/tabs/WebsiteTab.java
diff --git a/net/minecraft/launcher/ui/tabs/website/Browser.java b/src/main/java/net/minecraft/launcher/ui/tabs/website/Browser.java
similarity index 100%
rename from net/minecraft/launcher/ui/tabs/website/Browser.java
rename to src/main/java/net/minecraft/launcher/ui/tabs/website/Browser.java
diff --git a/net/minecraft/launcher/ui/tabs/website/JFXBrowser.java b/src/main/java/net/minecraft/launcher/ui/tabs/website/JFXBrowser.java
similarity index 100%
rename from net/minecraft/launcher/ui/tabs/website/JFXBrowser.java
rename to src/main/java/net/minecraft/launcher/ui/tabs/website/JFXBrowser.java
diff --git a/net/minecraft/launcher/ui/tabs/website/LegacySwingBrowser.java b/src/main/java/net/minecraft/launcher/ui/tabs/website/LegacySwingBrowser.java
similarity index 100%
rename from net/minecraft/launcher/ui/tabs/website/LegacySwingBrowser.java
rename to src/main/java/net/minecraft/launcher/ui/tabs/website/LegacySwingBrowser.java
diff --git a/net/minecraft/launcher/updater/AbstractDownloadInfo.java b/src/main/java/net/minecraft/launcher/updater/AbstractDownloadInfo.java
similarity index 100%
rename from net/minecraft/launcher/updater/AbstractDownloadInfo.java
rename to src/main/java/net/minecraft/launcher/updater/AbstractDownloadInfo.java
diff --git a/net/minecraft/launcher/updater/Argument.java b/src/main/java/net/minecraft/launcher/updater/Argument.java
similarity index 100%
rename from net/minecraft/launcher/updater/Argument.java
rename to src/main/java/net/minecraft/launcher/updater/Argument.java
diff --git a/net/minecraft/launcher/updater/ArgumentType.java b/src/main/java/net/minecraft/launcher/updater/ArgumentType.java
similarity index 100%
rename from net/minecraft/launcher/updater/ArgumentType.java
rename to src/main/java/net/minecraft/launcher/updater/ArgumentType.java
diff --git a/net/minecraft/launcher/updater/AssetIndexInfo.java b/src/main/java/net/minecraft/launcher/updater/AssetIndexInfo.java
similarity index 100%
rename from net/minecraft/launcher/updater/AssetIndexInfo.java
rename to src/main/java/net/minecraft/launcher/updater/AssetIndexInfo.java
diff --git a/net/minecraft/launcher/updater/CompleteMinecraftVersion.java b/src/main/java/net/minecraft/launcher/updater/CompleteMinecraftVersion.java
similarity index 100%
rename from net/minecraft/launcher/updater/CompleteMinecraftVersion.java
rename to src/main/java/net/minecraft/launcher/updater/CompleteMinecraftVersion.java
diff --git a/net/minecraft/launcher/updater/DownloadInfo.java b/src/main/java/net/minecraft/launcher/updater/DownloadInfo.java
similarity index 100%
rename from net/minecraft/launcher/updater/DownloadInfo.java
rename to src/main/java/net/minecraft/launcher/updater/DownloadInfo.java
diff --git a/net/minecraft/launcher/updater/DownloadType.java b/src/main/java/net/minecraft/launcher/updater/DownloadType.java
similarity index 100%
rename from net/minecraft/launcher/updater/DownloadType.java
rename to src/main/java/net/minecraft/launcher/updater/DownloadType.java
diff --git a/net/minecraft/launcher/updater/FileBasedVersionList.java b/src/main/java/net/minecraft/launcher/updater/FileBasedVersionList.java
similarity index 100%
rename from net/minecraft/launcher/updater/FileBasedVersionList.java
rename to src/main/java/net/minecraft/launcher/updater/FileBasedVersionList.java
diff --git a/net/minecraft/launcher/updater/Library.java b/src/main/java/net/minecraft/launcher/updater/Library.java
similarity index 100%
rename from net/minecraft/launcher/updater/Library.java
rename to src/main/java/net/minecraft/launcher/updater/Library.java
diff --git a/net/minecraft/launcher/updater/LibraryDownloadInfo.java b/src/main/java/net/minecraft/launcher/updater/LibraryDownloadInfo.java
similarity index 100%
rename from net/minecraft/launcher/updater/LibraryDownloadInfo.java
rename to src/main/java/net/minecraft/launcher/updater/LibraryDownloadInfo.java
diff --git a/net/minecraft/launcher/updater/LocalDownloadInfo.java b/src/main/java/net/minecraft/launcher/updater/LocalDownloadInfo.java
similarity index 100%
rename from net/minecraft/launcher/updater/LocalDownloadInfo.java
rename to src/main/java/net/minecraft/launcher/updater/LocalDownloadInfo.java
diff --git a/net/minecraft/launcher/updater/LocalVersionList.java b/src/main/java/net/minecraft/launcher/updater/LocalVersionList.java
similarity index 100%
rename from net/minecraft/launcher/updater/LocalVersionList.java
rename to src/main/java/net/minecraft/launcher/updater/LocalVersionList.java
diff --git a/net/minecraft/launcher/updater/MinecraftVersionManager.java b/src/main/java/net/minecraft/launcher/updater/MinecraftVersionManager.java
similarity index 100%
rename from net/minecraft/launcher/updater/MinecraftVersionManager.java
rename to src/main/java/net/minecraft/launcher/updater/MinecraftVersionManager.java
diff --git a/net/minecraft/launcher/updater/PartialVersion.java b/src/main/java/net/minecraft/launcher/updater/PartialVersion.java
similarity index 100%
rename from net/minecraft/launcher/updater/PartialVersion.java
rename to src/main/java/net/minecraft/launcher/updater/PartialVersion.java
diff --git a/net/minecraft/launcher/updater/PreHashedDownloadable.java b/src/main/java/net/minecraft/launcher/updater/PreHashedDownloadable.java
similarity index 100%
rename from net/minecraft/launcher/updater/PreHashedDownloadable.java
rename to src/main/java/net/minecraft/launcher/updater/PreHashedDownloadable.java
diff --git a/net/minecraft/launcher/updater/RemoteVersionList.java b/src/main/java/net/minecraft/launcher/updater/RemoteVersionList.java
similarity index 100%
rename from net/minecraft/launcher/updater/RemoteVersionList.java
rename to src/main/java/net/minecraft/launcher/updater/RemoteVersionList.java
diff --git a/net/minecraft/launcher/updater/VersionList.java b/src/main/java/net/minecraft/launcher/updater/VersionList.java
similarity index 100%
rename from net/minecraft/launcher/updater/VersionList.java
rename to src/main/java/net/minecraft/launcher/updater/VersionList.java
diff --git a/META-INF/log4j-provider.properties b/src/main/resources/META-INF/log4j-provider.properties
similarity index 100%
rename from META-INF/log4j-provider.properties
rename to src/main/resources/META-INF/log4j-provider.properties
diff --git a/dirt.png b/src/main/resources/dirt.png
similarity index 100%
rename from dirt.png
rename to src/main/resources/dirt.png
diff --git a/favicon.png b/src/main/resources/favicon.png
similarity index 100%
rename from favicon.png
rename to src/main/resources/favicon.png
diff --git a/launcher_properties.json b/src/main/resources/launcher_properties.json
similarity index 100%
rename from launcher_properties.json
rename to src/main/resources/launcher_properties.json
diff --git a/log4j2.xml b/src/main/resources/log4j2.xml
similarity index 100%
rename from log4j2.xml
rename to src/main/resources/log4j2.xml
diff --git a/minecraft_logo.png b/src/main/resources/minecraft_logo.png
similarity index 100%
rename from minecraft_logo.png
rename to src/main/resources/minecraft_logo.png
diff --git a/yggdrasil_session_pubkey.der b/src/main/resources/yggdrasil_session_pubkey.der
similarity index 100%
rename from yggdrasil_session_pubkey.der
rename to src/main/resources/yggdrasil_session_pubkey.der
-- 
2.35.1

