From 1089aedcdda2cc382e695a7013c6f8279022ec47 Mon Sep 17 00:00:00 2001
From: bigfoot547 <bigfoot@figboot.dev>
Date: Thu, 17 Mar 2022 16:58:53 -0500
Subject: [PATCH] Add xuid parameter and remove unused imports


diff --git a/src/main/java/com/mojang/launcher/game/process/GameProcessBuilder.java b/src/main/java/com/mojang/launcher/game/process/GameProcessBuilder.java
index c5df30c..dd3dd57 100644
--- a/src/main/java/com/mojang/launcher/game/process/GameProcessBuilder.java
+++ b/src/main/java/com/mojang/launcher/game/process/GameProcessBuilder.java
@@ -1,7 +1,6 @@
 package com.mojang.launcher.game.process;
 
 import com.google.common.base.MoreObjects;
-import com.google.common.base.Objects;
 import com.google.common.base.Predicate;
 import com.google.common.base.Predicates;
 import com.google.common.collect.Lists;
diff --git a/src/main/java/com/mojang/launcher/game/process/direct/DirectGameProcess.java b/src/main/java/com/mojang/launcher/game/process/direct/DirectGameProcess.java
index 2d0aca3..1b40430 100644
--- a/src/main/java/com/mojang/launcher/game/process/direct/DirectGameProcess.java
+++ b/src/main/java/com/mojang/launcher/game/process/direct/DirectGameProcess.java
@@ -1,7 +1,6 @@
 package com.mojang.launcher.game.process.direct;
 
 import com.google.common.base.MoreObjects;
-import com.google.common.base.Objects;
 import com.google.common.base.Predicate;
 import com.google.common.collect.EvictingQueue;
 import com.mojang.launcher.events.GameOutputLogProcessor;
diff --git a/src/main/java/com/mojang/util/QueueLogAppender.java b/src/main/java/com/mojang/util/QueueLogAppender.java
index 1abc840..8928e0c 100644
--- a/src/main/java/com/mojang/util/QueueLogAppender.java
+++ b/src/main/java/com/mojang/util/QueueLogAppender.java
@@ -11,13 +11,11 @@ import org.apache.logging.log4j.core.Filter;
 import org.apache.logging.log4j.core.Layout;
 import org.apache.logging.log4j.core.LogEvent;
 import org.apache.logging.log4j.core.appender.AbstractAppender;
-import org.apache.logging.log4j.core.config.Configuration;
 import org.apache.logging.log4j.core.config.plugins.Plugin;
 import org.apache.logging.log4j.core.config.plugins.PluginAttribute;
 import org.apache.logging.log4j.core.config.plugins.PluginElement;
 import org.apache.logging.log4j.core.config.plugins.PluginFactory;
 import org.apache.logging.log4j.core.layout.PatternLayout;
-import org.apache.logging.log4j.core.pattern.RegexReplacement;
 
 @Plugin(
     name = "Queue",
diff --git a/src/main/java/dev/figboot/olauncher/auth/MSLoginServer.java b/src/main/java/dev/figboot/olauncher/auth/MSLoginServer.java
index fa4c3cc..bb9cb51 100644
--- a/src/main/java/dev/figboot/olauncher/auth/MSLoginServer.java
+++ b/src/main/java/dev/figboot/olauncher/auth/MSLoginServer.java
@@ -12,7 +12,6 @@ import java.io.InputStream;
 import java.io.InputStreamReader;
 import java.net.*;
 import java.nio.charset.StandardCharsets;
-import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Consumer;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
diff --git a/src/main/java/dev/figboot/olauncher/auth/MicrosoftUserAuthentication.java b/src/main/java/dev/figboot/olauncher/auth/MicrosoftUserAuthentication.java
index 00a0675..3b2bd72 100644
--- a/src/main/java/dev/figboot/olauncher/auth/MicrosoftUserAuthentication.java
+++ b/src/main/java/dev/figboot/olauncher/auth/MicrosoftUserAuthentication.java
@@ -1,5 +1,8 @@
 package dev.figboot.olauncher.auth;
 
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
 import com.mojang.authlib.*;
 import com.mojang.authlib.exceptions.AuthenticationException;
 import com.mojang.authlib.minecraft.MinecraftSessionService;
@@ -11,12 +14,15 @@ import dev.figboot.olauncher.auth.responses.MSTokenAuthResponse;
 import dev.figboot.olauncher.auth.responses.MinecraftXboxLoginResponse;
 import dev.figboot.olauncher.auth.responses.XSTSAuthResponse;
 import dev.figboot.olauncher.auth.responses.XboxLiveAuthResponse;
+import lombok.Getter;
 import org.apache.commons.lang3.StringUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import java.nio.charset.StandardCharsets;
 import java.time.Instant;
 import java.time.temporal.ChronoUnit;
+import java.util.Base64;
 import java.util.Map;
 
 public class MicrosoftUserAuthentication extends HttpUserAuthentication {
@@ -31,6 +37,7 @@ public class MicrosoftUserAuthentication extends HttpUserAuthentication {
     private String xblToken;
     private Instant xblTokenExpire;
     private String mojToken;
+    @Getter private String xuid;
     private Instant mojTokenExpire;
     private boolean loggedIn;
 
@@ -161,6 +168,12 @@ public class MicrosoftUserAuthentication extends HttpUserAuthentication {
                 MinecraftXboxLoginResponse.class);
         this.mojToken = mcLoginRes.getAccessToken();
         this.mojTokenExpire = Instant.now().plus(mcLoginRes.getExpiresIn(), ChronoUnit.SECONDS);
+
+        // Extract xuid from mojToken (JWT)
+
+        JsonElement elem = JsonParser.parseString(new String(Base64.getUrlDecoder().decode(mojToken.split("\\.")[1].getBytes(StandardCharsets.UTF_8)), StandardCharsets.UTF_8));
+        JsonObject root = elem.getAsJsonObject();
+        this.xuid = root.get("xuid").getAsString();
     }
 
     private void profileLogIn() throws AuthenticationException {
@@ -232,6 +245,7 @@ public class MicrosoftUserAuthentication extends HttpUserAuthentication {
         msRefreshToken = (String)var1.get("refreshToken");
         xblToken = (String)var1.get("xboxLiveToken");
         xblTokenExpire = Instant.parse((String)var1.get("xboxLiveTokenExpire"));
+        xuid = (String)var1.get("xuid");
         mojToken = (String)var1.get("mojToken");
         mojTokenExpire = Instant.parse((String)var1.get("mojTokenExpire"));
     }
@@ -242,6 +256,7 @@ public class MicrosoftUserAuthentication extends HttpUserAuthentication {
         map.put("refreshToken", msRefreshToken);
         map.put("xboxLiveToken", xblToken);
         map.put("xboxLiveTokenExpire", xblTokenExpire.toString());
+        map.put("xuid", xuid);
         map.put("mojToken", mojToken);
         map.put("mojTokenExpire", mojTokenExpire.toString());
         return map;
diff --git a/src/main/java/dev/figboot/olauncher/auth/ui/MSLoginDialog.java b/src/main/java/dev/figboot/olauncher/auth/ui/MSLoginDialog.java
index 96f8b4a..f8b2ac1 100644
--- a/src/main/java/dev/figboot/olauncher/auth/ui/MSLoginDialog.java
+++ b/src/main/java/dev/figboot/olauncher/auth/ui/MSLoginDialog.java
@@ -3,7 +3,6 @@ package dev.figboot.olauncher.auth.ui;
 import com.mojang.launcher.OperatingSystem;
 import dev.figboot.olauncher.OLauncherConstants;
 import dev.figboot.olauncher.auth.MSLoginServer;
-import dev.figboot.olauncher.auth.MicrosoftUserAuthentication;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
diff --git a/src/main/java/net/minecraft/launcher/LauncherConstants.java b/src/main/java/net/minecraft/launcher/LauncherConstants.java
index c739732..db78fda 100644
--- a/src/main/java/net/minecraft/launcher/LauncherConstants.java
+++ b/src/main/java/net/minecraft/launcher/LauncherConstants.java
@@ -1,7 +1,6 @@
 package net.minecraft.launcher;
 
 import com.google.common.base.MoreObjects;
-import com.google.common.base.Objects;
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
 import com.mojang.launcher.updater.LowerCaseEnumTypeAdapterFactory;
diff --git a/src/main/java/net/minecraft/launcher/game/GameLaunchDispatcher.java b/src/main/java/net/minecraft/launcher/game/GameLaunchDispatcher.java
index c5a8fc4..a4b4e50 100644
--- a/src/main/java/net/minecraft/launcher/game/GameLaunchDispatcher.java
+++ b/src/main/java/net/minecraft/launcher/game/GameLaunchDispatcher.java
@@ -1,7 +1,6 @@
 package net.minecraft.launcher.game;
 
 import com.google.common.base.MoreObjects;
-import com.google.common.base.Objects;
 import com.google.common.collect.BiMap;
 import com.google.common.collect.HashBiMap;
 import com.mojang.authlib.UserAuthentication;
diff --git a/src/main/java/net/minecraft/launcher/game/MinecraftGameRunner.java b/src/main/java/net/minecraft/launcher/game/MinecraftGameRunner.java
index e82aa38..6f7050d 100644
--- a/src/main/java/net/minecraft/launcher/game/MinecraftGameRunner.java
+++ b/src/main/java/net/minecraft/launcher/game/MinecraftGameRunner.java
@@ -1,7 +1,6 @@
 package net.minecraft.launcher.game;
 
 import com.google.common.base.MoreObjects;
-import com.google.common.base.Objects;
 import com.google.common.base.Predicate;
 import com.google.gson.Gson;
 import com.google.gson.GsonBuilder;
@@ -46,6 +45,8 @@ import java.util.UUID;
 import java.util.Map.Entry;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipFile;
+
+import dev.figboot.olauncher.auth.MicrosoftUserAuthentication;
 import net.minecraft.launcher.CompatibilityRule;
 import net.minecraft.launcher.CurrentLaunchFeatureMatcher;
 import net.minecraft.launcher.Launcher;
@@ -318,6 +319,13 @@ public class MinecraftGameRunner extends AbstractGameRunner implements GameProce
         var6.put("classpath", this.constructClassPath(this.getVersion()));
         var6.put("classpath_separator", System.getProperty("path.separator"));
         var6.put("primary_jar", (new File(this.getLauncher().getWorkingDirectory(), "versions/" + this.getVersion().getJar() + "/" + this.getVersion().getJar() + ".jar")).getAbsolutePath());
+
+        // olauncher - add MSA options
+        if (var5 instanceof MicrosoftUserAuthentication) {
+            var6.put("auth_xuid", ((MicrosoftUserAuthentication)var5).getXuid());
+            // TODO: clientid as well
+        }
+
         return new StrSubstitutor(var6);
     }
 
diff --git a/src/main/java/net/minecraft/launcher/profile/Profile.java b/src/main/java/net/minecraft/launcher/profile/Profile.java
index a450ca3..fdd24fc 100644
--- a/src/main/java/net/minecraft/launcher/profile/Profile.java
+++ b/src/main/java/net/minecraft/launcher/profile/Profile.java
@@ -1,7 +1,6 @@
 package net.minecraft.launcher.profile;
 
 import com.google.common.base.MoreObjects;
-import com.google.common.base.Objects;
 import com.google.common.collect.Sets;
 import com.mojang.launcher.updater.VersionFilter;
 import com.mojang.launcher.versions.ReleaseType;
diff --git a/src/main/java/net/minecraft/launcher/updater/CompleteMinecraftVersion.java b/src/main/java/net/minecraft/launcher/updater/CompleteMinecraftVersion.java
index b9ac4df..9d786d5 100644
--- a/src/main/java/net/minecraft/launcher/updater/CompleteMinecraftVersion.java
+++ b/src/main/java/net/minecraft/launcher/updater/CompleteMinecraftVersion.java
@@ -1,7 +1,6 @@
 package net.minecraft.launcher.updater;
 
 import com.google.common.base.MoreObjects;
-import com.google.common.base.Objects;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
 import com.google.common.collect.Sets;
-- 
2.47.2

