From 78d7f1a029ac3d407f29676f5583864871d8d1ff Mon Sep 17 00:00:00 2001
From: bigfoot547 <bigfoot@figboot.dev>
Date: Sat, 26 Mar 2022 15:24:44 -0500
Subject: [PATCH] Add bundled java runtime functionality


diff --git a/src/main/java/com/mojang/launcher/Launcher.java b/src/main/java/com/mojang/launcher/Launcher.java
index f79a4b6..100ab0f 100644
--- a/src/main/java/com/mojang/launcher/Launcher.java
+++ b/src/main/java/com/mojang/launcher/Launcher.java
@@ -9,6 +9,10 @@ import java.net.PasswordAuthentication;
 import java.net.Proxy;
 import java.util.concurrent.ThreadPoolExecutor;
 import java.util.concurrent.TimeUnit;
+
+import dev.figboot.olauncher.OLauncherConstants;
+import dev.figboot.olauncher.launcher.runtime.JavaRuntimeManager;
+import lombok.Getter;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -24,6 +28,8 @@ public class Launcher {
     private final ReleaseTypeFactory releaseTypeFactory;
     private final int launcherFormatVersion;
 
+    @Getter private final JavaRuntimeManager jreManager; // olauncher - add java runtime manager
+
     public Launcher(UserInterface var1, File var2, Proxy var3, PasswordAuthentication var4, VersionManager var5, Agent var6, ReleaseTypeFactory var7, int var8) {
         this.downloaderExecutorService = new ExceptionalThreadPoolExecutor(16, 16, 30L, TimeUnit.SECONDS);
         this.ui = var1;
@@ -35,6 +41,8 @@ public class Launcher {
         this.releaseTypeFactory = var7;
         this.launcherFormatVersion = var8;
         this.downloaderExecutorService.allowCoreThreadTimeOut(true);
+
+        this.jreManager = new JavaRuntimeManager(OLauncherConstants.JRE_MANIFEST_URL, var3, var2); // olauncher - add java runtime manager
     }
 
     public ReleaseTypeFactory getReleaseTypeFactory() {
diff --git a/src/main/java/com/mojang/launcher/game/runner/AbstractGameRunner.java b/src/main/java/com/mojang/launcher/game/runner/AbstractGameRunner.java
index f3546fc..c8cc04d 100644
--- a/src/main/java/com/mojang/launcher/game/runner/AbstractGameRunner.java
+++ b/src/main/java/com/mojang/launcher/game/runner/AbstractGameRunner.java
@@ -13,6 +13,8 @@ import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
+
+import dev.figboot.olauncher.launcher.runtime.RuntimesManifest;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
@@ -115,6 +117,10 @@ public abstract class AbstractGameRunner implements GameRunner, DownloadListener
         }
     }
 
+    private void downloadRuntime() {
+
+    }
+
     protected void downloadRequiredFiles(VersionSyncInfo var1) {
         try {
             DownloadJob var2 = new DownloadJob("Version & Libraries", false, this);
diff --git a/src/main/java/dev/figboot/olauncher/OLauncherConstants.java b/src/main/java/dev/figboot/olauncher/OLauncherConstants.java
index 67728f6..e88d992 100644
--- a/src/main/java/dev/figboot/olauncher/OLauncherConstants.java
+++ b/src/main/java/dev/figboot/olauncher/OLauncherConstants.java
@@ -18,8 +18,11 @@
 
 package dev.figboot.olauncher;
 
+import java.net.URL;
 import java.util.UUID;
 
+import static net.minecraft.launcher.LauncherConstants.constantURL;
+
 public final class OLauncherConstants {
     private OLauncherConstants() { }
 
@@ -28,4 +31,6 @@ public final class OLauncherConstants {
     public static final String APP_SCOPES = "XboxLive.signin offline_access";
     public static final int REDIR_URI_PORT = 6183;
     public static final String REDIR_URI = "http://localhost:" + REDIR_URI_PORT;
+
+    public static final URL JRE_MANIFEST_URL = constantURL("https://launchermeta.mojang.com/v1/products/java-runtime/2ec0cc96c44e5a76b9c8b7c39df7210883d12871/all.json");
 }
diff --git a/src/main/java/dev/figboot/olauncher/launcher/JavaVersionInfo.java b/src/main/java/dev/figboot/olauncher/launcher/JavaVersionInfo.java
new file mode 100644
index 0000000..61c8411
--- /dev/null
+++ b/src/main/java/dev/figboot/olauncher/launcher/JavaVersionInfo.java
@@ -0,0 +1,33 @@
+/*
+ * OLauncher
+ * Copyright (C) 2022  bigfoot547 <olauncher@figboot.dev>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see https://github.com/olauncher/olauncher .
+ */
+
+package dev.figboot.olauncher.launcher;
+
+import lombok.Getter;
+
+public class JavaVersionInfo {
+    @Getter private String component;
+    @Getter private int majorVersion;
+
+    public JavaVersionInfo() { }
+
+    public JavaVersionInfo(JavaVersionInfo other) {
+        this.component = other.component;
+        this.majorVersion = other.majorVersion;
+    }
+}
diff --git a/src/main/java/dev/figboot/olauncher/launcher/runtime/JavaRuntimeManager.java b/src/main/java/dev/figboot/olauncher/launcher/runtime/JavaRuntimeManager.java
new file mode 100644
index 0000000..0595342
--- /dev/null
+++ b/src/main/java/dev/figboot/olauncher/launcher/runtime/JavaRuntimeManager.java
@@ -0,0 +1,219 @@
+/*
+ * OLauncher
+ * Copyright (C) 2022  bigfoot547 <olauncher@figboot.dev>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see https://github.com/olauncher/olauncher .
+ */
+
+package dev.figboot.olauncher.launcher.runtime;
+
+import com.google.common.reflect.TypeToken;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.mojang.launcher.Http;
+import com.mojang.launcher.OperatingSystem;
+import com.mojang.launcher.updater.ExceptionalThreadPoolExecutor;
+import com.mojang.launcher.updater.download.DownloadJob;
+import dev.figboot.olauncher.launcher.runtime.file.RuntimeFile;
+import dev.figboot.olauncher.launcher.runtime.file.RuntimeFileFile;
+import dev.figboot.olauncher.launcher.runtime.file.RuntimeFileList;
+import dev.figboot.olauncher.util.DeletingFileVisitor;
+import dev.figboot.olauncher.util.InstantTypeAdapter;
+import dev.figboot.olauncher.util.runtime.RuntimeFileDeserializer;
+import lombok.Getter;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.BufferedReader;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.math.BigInteger;
+import java.net.HttpURLConnection;
+import java.net.Proxy;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.DigestInputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.time.Instant;
+import java.util.*;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+public class JavaRuntimeManager {
+    private static final Logger LOGGER = LogManager.getLogger();
+
+    private boolean refreshing;
+    private final Object refreshingLock = new Object();
+
+    private final URL manifestUrl;
+    private final Proxy proxy;
+
+    private final Path runtimesDirectory;
+    private final Path platformDirectory;
+    private final Gson gson;
+
+    private final String jrePlatformName;
+    @Getter private final ThreadPoolExecutor executorService;
+    private Map<String, Map<String, List<RuntimesManifest.Runtime>>> manifest;
+
+    public JavaRuntimeManager(URL manifestUrl, Proxy proxy, File workDir) {
+        refreshing = false;
+        this.manifestUrl = manifestUrl;
+        this.proxy = proxy;
+
+        runtimesDirectory = workDir.toPath().resolve("olruntime");
+        gson = new GsonBuilder()
+                .registerTypeAdapter(Instant.class, new InstantTypeAdapter())
+                .registerTypeAdapter(RuntimeFile.class, new RuntimeFileDeserializer())
+                .create();
+
+        jrePlatformName = findPlatformName(System.getProperty("os.arch").contains("64")); // FIXME: make accurate
+        platformDirectory = runtimesDirectory.resolve(jrePlatformName);
+        executorService = new ExceptionalThreadPoolExecutor(4, 8, 30L, TimeUnit.SECONDS);
+    }
+
+    public void reloadRuntimes() throws IOException {
+        synchronized (refreshingLock) {
+            refreshing = true;
+        }
+
+        manifest = gson.fromJson(Http.performGet(manifestUrl, proxy), new TypeToken<Map<String, Map<String, List<RuntimesManifest.Runtime>>>>(){}.getType());
+
+        LOGGER.info("Your java runtime platform is '" + jrePlatformName + "'.");
+
+        if (!manifest.containsKey(jrePlatformName)) {
+            LOGGER.error("Your JVM platform (" + jrePlatformName + ") is NOT PRESENT in the runtime manifest! The launcher will now break spectacularly.");
+        }
+
+        LOGGER.info("Finished refreshing java runtime list.");
+        synchronized (refreshingLock) {
+            refreshing = false;
+        }
+    }
+
+    public List<RuntimesManifest.Runtime> getRuntimes(String component) {
+        synchronized (refreshingLock) {
+            if (refreshing) return Collections.emptyList();
+        }
+
+        List<RuntimesManifest.Runtime> runtimes = manifest.get(jrePlatformName).get(component);
+        if (runtimes == null) return Collections.emptyList();
+        return Collections.unmodifiableList(runtimes);
+    }
+
+    public boolean shouldDownloadRuntime(String component, RuntimesManifest.Runtime runtime) throws IOException {
+        Path componentFolder = platformDirectory.resolve(component);
+
+        if (!Files.isDirectory(componentFolder)) {
+            return true;
+        }
+
+        Path descPath = componentFolder.resolve(".version");
+        if (Files.notExists(descPath)) return true;
+
+        Instant updateTime;
+        String version;
+
+        try (DataInputStream dis = new DataInputStream(Files.newInputStream(descPath))) {
+            updateTime = Instant.ofEpochMilli(dis.readLong());
+            version = dis.readUTF();
+        }
+
+        return runtime.getVersion().getReleased().isAfter(updateTime) || runtime.getVersion().getName().equals(version);
+    }
+
+    public Path getRuntimeDirectory(String component) {
+        return platformDirectory.resolve(component);
+    }
+
+    public void deleteRuntime(String component) throws IOException {
+        Path componentDir = platformDirectory.resolve(component);
+
+        Files.walkFileTree(componentDir, DeletingFileVisitor.get());
+    }
+
+    public RuntimeFileList preDownloadRuntime(String component, RuntimesManifest.Runtime runtime) throws IOException {
+        Path componentDir = platformDirectory.resolve(component);
+        HttpURLConnection conn = (HttpURLConnection)runtime.getManifest().getUrl().openConnection();
+        if (conn.getResponseCode() / 100 != 2) throw new RuntimeException("Error downloading component manifest '" + component + "': Received response " + conn.getResponseCode() + " " + conn.getResponseMessage());
+
+        MessageDigest digest;
+        try {
+            digest = MessageDigest.getInstance("SHA-1");
+        } catch (NoSuchAlgorithmException ex) {
+            throw new RuntimeException(ex); // should not happen
+        }
+
+        RuntimeFileList files;
+        try (Reader reader = new BufferedReader(new InputStreamReader(new DigestInputStream(conn.getInputStream(), digest)))) {
+            files = gson.fromJson(reader, RuntimeFileList.class);
+        }
+
+        String hash = String.format("%1$040x", new BigInteger(1, digest.digest()));
+        if (!hash.equalsIgnoreCase(runtime.getManifest().getSha1()))
+            throw new RuntimeException("Error downloading component manifest '" + component + "': The manifest hash does not match! (Expected " + runtime.getManifest().getSha1() + ", got " + hash + ")");
+
+        Path runtimeRootDir = componentDir.resolve("runtime");
+
+        for (Map.Entry<String, RuntimeFile> entry : files.getFiles().entrySet()) {
+            if (entry.getValue() instanceof RuntimeFileFile) continue;
+            entry.getValue().create(runtimeRootDir.resolve(entry.getKey()));
+        }
+
+        return files;
+    }
+
+    public void postDownloadRuntime(String component, RuntimesManifest.Runtime runtime) throws IOException {
+        Path componentFolder = platformDirectory.resolve(component);
+
+        // Write .version
+        Path descPath = componentFolder.resolve(".version");
+        if (Files.notExists(descPath)) {
+            try (DataOutputStream dos = new DataOutputStream(Files.newOutputStream(descPath))) {
+                dos.writeLong(runtime.getVersion().getReleased().toEpochMilli());
+                dos.writeUTF(runtime.getVersion().getName());
+            }
+        }
+    }
+
+    public DownloadJob downloadRuntime(DownloadJob job, String component, RuntimesManifest.Runtime runtime, RuntimeFileList fileList) throws IOException {
+        RuntimeFileFile file;
+        for (Map.Entry<String, RuntimeFile> entry : fileList.getFiles().entrySet()) {
+            if (!(entry.getValue() instanceof RuntimeFileFile)) continue;
+            file = (RuntimeFileFile)entry.getValue();
+            // TODO: add thing
+        }
+        return job;
+    }
+
+    private static String findPlatformName(boolean _64bit) {
+        switch (OperatingSystem.getCurrentPlatform()) {
+            case WINDOWS:
+                return _64bit ? "windows-x64" : "windows-x86";
+            case OSX:
+                return "mac-os";
+            case LINUX:
+                return _64bit ? "linux" : "linux-i386"; // not technically correct
+            default:
+                return "gamecore";
+        }
+    }
+}
diff --git a/src/main/java/dev/figboot/olauncher/launcher/runtime/RuntimesManifest.java b/src/main/java/dev/figboot/olauncher/launcher/runtime/RuntimesManifest.java
new file mode 100644
index 0000000..a0bc907
--- /dev/null
+++ b/src/main/java/dev/figboot/olauncher/launcher/runtime/RuntimesManifest.java
@@ -0,0 +1,38 @@
+/*
+ * OLauncher
+ * Copyright (C) 2022  bigfoot547 <olauncher@figboot.dev>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see https://github.com/olauncher/olauncher .
+ */
+
+package dev.figboot.olauncher.launcher.runtime;
+
+import lombok.Getter;
+import net.minecraft.launcher.updater.DownloadInfo;
+
+import java.time.Instant;
+
+public final class RuntimesManifest {
+    private RuntimesManifest() { }
+
+    public static class Runtime {
+        @Getter private DownloadInfo manifest;
+        @Getter private Version version;
+    }
+
+    public static class Version {
+        @Getter private String name;
+        @Getter private Instant released;
+    }
+}
diff --git a/src/main/java/dev/figboot/olauncher/launcher/runtime/download/RuntimeFileDownloadable.java b/src/main/java/dev/figboot/olauncher/launcher/runtime/download/RuntimeFileDownloadable.java
new file mode 100644
index 0000000..cf30719
--- /dev/null
+++ b/src/main/java/dev/figboot/olauncher/launcher/runtime/download/RuntimeFileDownloadable.java
@@ -0,0 +1,23 @@
+package dev.figboot.olauncher.launcher.runtime.download;
+
+import com.mojang.launcher.updater.download.Downloadable;
+import dev.figboot.olauncher.launcher.runtime.file.RuntimeFileFile;
+
+import java.io.IOException;
+import java.net.Proxy;
+import java.nio.file.Path;
+
+public class RuntimeFileDownloadable extends Downloadable {
+    private final RuntimeFileFile file;
+
+    public RuntimeFileDownloadable(Proxy proxy, Path finalLoc, RuntimeFileFile file) {
+        super(proxy, file.getRaw().getUrl(), finalLoc.toFile(), false);
+        this.file = file;
+    }
+
+    @Override
+    public String download() throws IOException {
+
+        return null;
+    }
+}
diff --git a/src/main/java/dev/figboot/olauncher/launcher/runtime/file/RuntimeFile.java b/src/main/java/dev/figboot/olauncher/launcher/runtime/file/RuntimeFile.java
new file mode 100644
index 0000000..1d4fdae
--- /dev/null
+++ b/src/main/java/dev/figboot/olauncher/launcher/runtime/file/RuntimeFile.java
@@ -0,0 +1,26 @@
+/*
+ * OLauncher
+ * Copyright (C) 2022  bigfoot547 <olauncher@figboot.dev>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see https://github.com/olauncher/olauncher .
+ */
+
+package dev.figboot.olauncher.launcher.runtime.file;
+
+import java.io.IOException;
+import java.nio.file.Path;
+
+public interface RuntimeFile {
+    void create(Path path) throws IOException;
+}
diff --git a/src/main/java/dev/figboot/olauncher/launcher/runtime/file/RuntimeFileDirectory.java b/src/main/java/dev/figboot/olauncher/launcher/runtime/file/RuntimeFileDirectory.java
new file mode 100644
index 0000000..756a5c7
--- /dev/null
+++ b/src/main/java/dev/figboot/olauncher/launcher/runtime/file/RuntimeFileDirectory.java
@@ -0,0 +1,30 @@
+/*
+ * OLauncher
+ * Copyright (C) 2022  bigfoot547 <olauncher@figboot.dev>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see https://github.com/olauncher/olauncher .
+ */
+
+package dev.figboot.olauncher.launcher.runtime.file;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+public class RuntimeFileDirectory implements RuntimeFile {
+    @Override
+    public void create(Path path) throws IOException {
+        Files.createDirectories(path);
+    }
+}
diff --git a/src/main/java/dev/figboot/olauncher/launcher/runtime/file/RuntimeFileFile.java b/src/main/java/dev/figboot/olauncher/launcher/runtime/file/RuntimeFileFile.java
new file mode 100644
index 0000000..627034c
--- /dev/null
+++ b/src/main/java/dev/figboot/olauncher/launcher/runtime/file/RuntimeFileFile.java
@@ -0,0 +1,35 @@
+/*
+ * OLauncher
+ * Copyright (C) 2022  bigfoot547 <olauncher@figboot.dev>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see https://github.com/olauncher/olauncher .
+ */
+
+package dev.figboot.olauncher.launcher.runtime.file;
+
+import lombok.Getter;
+import net.minecraft.launcher.updater.DownloadInfo;
+
+import java.io.IOException;
+import java.nio.file.Path;
+
+public class RuntimeFileFile implements RuntimeFile {
+    @Getter private boolean executable;
+    @Getter private DownloadInfo raw, lzma;
+
+    @Override
+    public void create(Path path) throws IOException {
+        throw new UnsupportedOperationException("The file must be downloaded");
+    }
+}
diff --git a/src/main/java/dev/figboot/olauncher/launcher/runtime/file/RuntimeFileLink.java b/src/main/java/dev/figboot/olauncher/launcher/runtime/file/RuntimeFileLink.java
new file mode 100644
index 0000000..880ad87
--- /dev/null
+++ b/src/main/java/dev/figboot/olauncher/launcher/runtime/file/RuntimeFileLink.java
@@ -0,0 +1,35 @@
+/*
+ * OLauncher
+ * Copyright (C) 2022  bigfoot547 <olauncher@figboot.dev>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see https://github.com/olauncher/olauncher .
+ */
+
+package dev.figboot.olauncher.launcher.runtime.file;
+
+import lombok.Getter;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+public class RuntimeFileLink implements RuntimeFile {
+    @Getter private String target;
+
+    @Override
+    public void create(Path path) throws IOException {
+        Files.createSymbolicLink(path, Paths.get(target));
+    }
+}
diff --git a/src/main/java/dev/figboot/olauncher/launcher/runtime/file/RuntimeFileList.java b/src/main/java/dev/figboot/olauncher/launcher/runtime/file/RuntimeFileList.java
new file mode 100644
index 0000000..3d8f688
--- /dev/null
+++ b/src/main/java/dev/figboot/olauncher/launcher/runtime/file/RuntimeFileList.java
@@ -0,0 +1,27 @@
+/*
+ * OLauncher
+ * Copyright (C) 2022  bigfoot547 <olauncher@figboot.dev>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see https://github.com/olauncher/olauncher .
+ */
+
+package dev.figboot.olauncher.launcher.runtime.file;
+
+import lombok.Getter;
+
+import java.util.Map;
+
+public class RuntimeFileList {
+    @Getter private Map<String, RuntimeFile> files;
+}
diff --git a/src/main/java/dev/figboot/olauncher/util/DeletingFileVisitor.java b/src/main/java/dev/figboot/olauncher/util/DeletingFileVisitor.java
new file mode 100644
index 0000000..c35c40d
--- /dev/null
+++ b/src/main/java/dev/figboot/olauncher/util/DeletingFileVisitor.java
@@ -0,0 +1,59 @@
+/*
+ * OLauncher
+ * Copyright (C) 2022  bigfoot547 <olauncher@figboot.dev>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see https://github.com/olauncher/olauncher .
+ */
+
+package dev.figboot.olauncher.util;
+
+import java.io.IOException;
+import java.nio.file.FileVisitResult;
+import java.nio.file.FileVisitor;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.BasicFileAttributes;
+
+public class DeletingFileVisitor implements FileVisitor<Path> {
+    private static final DeletingFileVisitor INSTANCE = new DeletingFileVisitor();
+
+    public static DeletingFileVisitor get() {
+        return INSTANCE;
+    }
+
+    private DeletingFileVisitor() { }
+
+    @Override
+    public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) {
+        return FileVisitResult.CONTINUE;
+    }
+
+    @Override
+    public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+        Files.delete(file);
+        return FileVisitResult.CONTINUE;
+    }
+
+    @Override
+    public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
+        throw exc;
+    }
+
+    @Override
+    public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
+        if (exc != null) throw exc;
+        Files.delete(dir);
+        return FileVisitResult.CONTINUE;
+    }
+}
diff --git a/src/main/java/dev/figboot/olauncher/util/runtime/RuntimeFileDeserializer.java b/src/main/java/dev/figboot/olauncher/util/runtime/RuntimeFileDeserializer.java
new file mode 100644
index 0000000..3499f3d
--- /dev/null
+++ b/src/main/java/dev/figboot/olauncher/util/runtime/RuntimeFileDeserializer.java
@@ -0,0 +1,46 @@
+/*
+ * OLauncher
+ * Copyright (C) 2022  bigfoot547 <olauncher@figboot.dev>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see https://github.com/olauncher/olauncher .
+ */
+
+package dev.figboot.olauncher.util.runtime;
+
+import com.google.gson.*;
+import dev.figboot.olauncher.launcher.runtime.file.RuntimeFile;
+import dev.figboot.olauncher.launcher.runtime.file.RuntimeFileDirectory;
+import dev.figboot.olauncher.launcher.runtime.file.RuntimeFileFile;
+import dev.figboot.olauncher.launcher.runtime.file.RuntimeFileLink;
+
+import java.lang.reflect.Type;
+
+public class RuntimeFileDeserializer implements JsonDeserializer<RuntimeFile> {
+    @Override
+    public RuntimeFile deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
+        JsonObject obj = json.getAsJsonObject();
+        String type = obj.get("type").getAsString();
+
+        switch (type) {
+            case "file":
+                return context.deserialize(obj, RuntimeFileFile.class);
+            case "directory":
+                return context.deserialize(obj, RuntimeFileDirectory.class);
+            case "link":
+                return context.deserialize(obj, RuntimeFileLink.class);
+        }
+
+        throw new JsonParseException("Unknown file type '" + type + "'");
+    }
+}
diff --git a/src/main/java/net/minecraft/launcher/Launcher.java b/src/main/java/net/minecraft/launcher/Launcher.java
index a9a6ee3..1ed2f8f 100644
--- a/src/main/java/net/minecraft/launcher/Launcher.java
+++ b/src/main/java/net/minecraft/launcher/Launcher.java
@@ -31,11 +31,15 @@ import java.util.List;
 import java.util.Locale;
 import java.util.UUID;
 import javax.swing.JFrame;
+
+import dev.figboot.olauncher.OLauncherConstants;
+import dev.figboot.olauncher.launcher.runtime.JavaRuntimeManager;
 import joptsimple.ArgumentAcceptingOptionSpec;
 import joptsimple.NonOptionArgumentSpec;
 import joptsimple.OptionException;
 import joptsimple.OptionParser;
 import joptsimple.OptionSet;
+import lombok.Getter;
 import net.minecraft.launcher.game.GameLaunchDispatcher;
 import net.minecraft.launcher.game.MinecraftReleaseType;
 import net.minecraft.launcher.game.MinecraftReleaseTypeFactory;
@@ -110,6 +114,7 @@ public class Launcher {
             this.launcher = new com.mojang.launcher.Launcher(this.userInterface, var2, var3, var4, new MinecraftVersionManager(new LocalVersionList(var2), new RemoteVersionList(LauncherConstants.PROPERTIES.getVersionManifest(), var3)), Agent.MINECRAFT, MinecraftReleaseTypeFactory.instance(), 21);
             this.profileManager = new ProfileManager(this);
             ((SwingUserInterface)this.userInterface).initializeFrame();
+            refreshJavaRuntimes(); // olauncher - refresh java runtimes
             this.refreshVersionsAndProfiles();
         }
     }
@@ -171,6 +176,18 @@ public class Launcher {
         return (String[])var6.toArray(new String[var6.size()]);
     }
 
+    // olauncher start - refresh java runtime function
+    public void refreshJavaRuntimes() {
+        getLauncher().getJreManager().getExecutorService().submit(() -> {
+            try {
+                getLauncher().getJreManager().reloadRuntimes();
+            } catch (Throwable t) {
+                Launcher.LOGGER.error("Unexpected exception refreshing java runtime list", t);
+            }
+        });
+    }
+    // olauncher end
+
     public void refreshVersionsAndProfiles() {
         this.getLauncher().getVersionManager().getExecutorService().submit(new Runnable() {
             public void run() {
diff --git a/src/main/java/net/minecraft/launcher/SwingUserInterface.java b/src/main/java/net/minecraft/launcher/SwingUserInterface.java
index 06e506a..a7942e5 100644
--- a/src/main/java/net/minecraft/launcher/SwingUserInterface.java
+++ b/src/main/java/net/minecraft/launcher/SwingUserInterface.java
@@ -103,6 +103,7 @@ public class SwingUserInterface implements MinecraftUserInterface {
                 SwingUserInterface.this.frame.dispose();
                 SwingUserInterface.LOGGER.info("Halting executors");
                 SwingUserInterface.this.minecraftLauncher.getLauncher().getVersionManager().getExecutorService().shutdown();
+                SwingUserInterface.this.minecraftLauncher.getLauncher().getJreManager().getExecutorService().shutdown(); // olauncher - shut down jreManager executor
                 SwingUserInterface.LOGGER.info("Awaiting termination.");
 
                 try {
diff --git a/src/main/java/net/minecraft/launcher/game/MinecraftGameRunner.java b/src/main/java/net/minecraft/launcher/game/MinecraftGameRunner.java
index 6f7050d..37f787d 100644
--- a/src/main/java/net/minecraft/launcher/game/MinecraftGameRunner.java
+++ b/src/main/java/net/minecraft/launcher/game/MinecraftGameRunner.java
@@ -34,6 +34,7 @@ import java.io.InputStreamReader;
 import java.net.InetSocketAddress;
 import java.net.PasswordAuthentication;
 import java.net.Proxy;
+import java.nio.file.Files;
 import java.util.Collection;
 import java.util.Date;
 import java.util.Enumeration;
@@ -47,6 +48,7 @@ import java.util.zip.ZipEntry;
 import java.util.zip.ZipFile;
 
 import dev.figboot.olauncher.auth.MicrosoftUserAuthentication;
+import dev.figboot.olauncher.launcher.runtime.RuntimesManifest;
 import net.minecraft.launcher.CompatibilityRule;
 import net.minecraft.launcher.CurrentLaunchFeatureMatcher;
 import net.minecraft.launcher.Launcher;
diff --git a/src/main/java/net/minecraft/launcher/updater/CompleteMinecraftVersion.java b/src/main/java/net/minecraft/launcher/updater/CompleteMinecraftVersion.java
index 9d786d5..3f95163 100644
--- a/src/main/java/net/minecraft/launcher/updater/CompleteMinecraftVersion.java
+++ b/src/main/java/net/minecraft/launcher/updater/CompleteMinecraftVersion.java
@@ -25,6 +25,9 @@ import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.Map.Entry;
+
+import dev.figboot.olauncher.launcher.JavaVersionInfo;
+import lombok.Getter;
 import net.minecraft.launcher.CompatibilityRule;
 import net.minecraft.launcher.CurrentLaunchFeatureMatcher;
 import net.minecraft.launcher.Launcher;
@@ -54,6 +57,8 @@ public class CompleteMinecraftVersion implements CompleteVersion {
     private AssetIndexInfo assetIndex;
     private Map<ArgumentType, List<Argument>> arguments;
 
+    @Getter private JavaVersionInfo javaVersion; // olauncher - add java version
+
     public CompleteMinecraftVersion() {
     }
 
@@ -105,6 +110,8 @@ public class CompleteMinecraftVersion implements CompleteVersion {
             }
         }
 
+        // olauncher - handle java version
+        this.javaVersion = new JavaVersionInfo(var1.javaVersion);
     }
 
     public String getId() {
@@ -362,6 +369,9 @@ public class CompleteMinecraftVersion implements CompleteVersion {
                 }
             }
 
+            // olauncher - handle java version
+            if (this.javaVersion != null) var5.javaVersion = new JavaVersionInfo(this.javaVersion);
+
             return var5;
         }
     }
-- 
2.35.1

