From 894d996ece13d6a368a2034c37ba8df31e8542c5 Mon Sep 17 00:00:00 2001
From: bigfoot547 <bigfoot3132@gmail.com>
Date: Sat, 10 Jun 2023 16:05:54 -0500
Subject: [PATCH] Allow to customize window title


diff --git a/src/main/java/net/minecraft/launcher/Main.java b/src/main/java/net/minecraft/launcher/Main.java
index fff6d93..6668dd0 100644
--- a/src/main/java/net/minecraft/launcher/Main.java
+++ b/src/main/java/net/minecraft/launcher/Main.java
@@ -56,7 +56,7 @@ public class Main {
         var10.mkdirs();
         LOGGER.debug("About to create JFrame.");
         JFrame var12 = new JFrame();
-        var12.setTitle(LauncherVersion.getLauncherName() + " " + LauncherVersion.getVersion() + LauncherConstants.PROPERTIES.getEnvironment().getTitle()); // olauncher - update version
+        var12.setTitle(SwingUserInterface.WINDOW_TITLE == null ? LauncherVersion.getLauncherName() + " " + LauncherVersion.getVersion() + LauncherConstants.PROPERTIES.getEnvironment().getTitle() : SwingUserInterface.WINDOW_TITLE); // olauncher - update version
         var12.setPreferredSize(new Dimension(900, 580));
 
         try {
diff --git a/src/main/java/net/minecraft/launcher/SwingUserInterface.java b/src/main/java/net/minecraft/launcher/SwingUserInterface.java
index 524c1f7..528aa6c 100644
--- a/src/main/java/net/minecraft/launcher/SwingUserInterface.java
+++ b/src/main/java/net/minecraft/launcher/SwingUserInterface.java
@@ -45,6 +45,8 @@ public class SwingUserInterface implements MinecraftUserInterface {
     private LauncherPanel launcherPanel;
     private final JFrame frame;
 
+    public static final String WINDOW_TITLE = System.getProperty("net.minecraft.launcher.WindowTitle"); // olauncher - window title
+
     public SwingUserInterface(Launcher var1, JFrame var2) {
         this.minecraftLauncher = var1;
         this.frame = var2;
@@ -95,7 +97,7 @@ public class SwingUserInterface implements MinecraftUserInterface {
 
     public void initializeFrame() {
         this.frame.getContentPane().removeAll();
-        this.frame.setTitle(LauncherVersion.getLauncherName() + " " + LauncherVersion.getVersion() + LauncherConstants.PROPERTIES.getEnvironment().getTitle()); // olauncher - update version
+        this.frame.setTitle(WINDOW_TITLE == null ? LauncherVersion.getLauncherName() + " " + LauncherVersion.getVersion() + LauncherConstants.PROPERTIES.getEnvironment().getTitle() : WINDOW_TITLE); // olauncher - update version
         this.frame.setPreferredSize(new Dimension(900, 580));
         this.frame.setDefaultCloseOperation(2);
         this.frame.addWindowListener(new WindowAdapter() {
-- 
2.40.1

