From 1961d86659782f0b5098e68162ee676f6015ab77 Mon Sep 17 00:00:00 2001
From: bigfoot547 <bigfoot3132@gmail.com>
Date: Sun, 31 Mar 2024 17:49:36 -0500
Subject: [PATCH] [WIP/DRAFT] Auth overhaul


diff --git a/src/main/java/dev/figboot/olauncher/auth/MicrosoftUserAuthentication.java b/src/main/java/dev/figboot/olauncher/auth/MicrosoftUserAuthentication.java
index b6c71f6..b6065e8 100644
--- a/src/main/java/dev/figboot/olauncher/auth/MicrosoftUserAuthentication.java
+++ b/src/main/java/dev/figboot/olauncher/auth/MicrosoftUserAuthentication.java
@@ -355,7 +355,7 @@ public class MicrosoftUserAuthentication extends HttpUserAuthentication {
     }
 
     private static UUID xuidFromString(String xuidStr) {
-        return new UUID(0, Long.parseUnsignedLong(xuidStr));
+        return StringUtils.isBlank(xuidStr) ? null : new UUID(0, Long.parseUnsignedLong(xuidStr));
     }
 
     @Override
@@ -372,7 +372,7 @@ public class MicrosoftUserAuthentication extends HttpUserAuthentication {
         }
 
         if (var1.containsKey("xuid")) {
-            xuid = xuidFromString("xuid");
+            xuid = xuidFromString((String)var1.get("xuid"));
         } else if (var1.containsKey("xuid2")) {
             try {
                 xuid = UUID.fromString((String)var1.get("xuid2"));
@@ -399,7 +399,11 @@ public class MicrosoftUserAuthentication extends HttpUserAuthentication {
         map.put("refreshToken", msRefreshToken);
         map.put("xboxLiveToken", xblToken);
         map.put("xboxLiveTokenExpire", xblTokenExpire.toString());
-        map.put("xuid2", xuid.toString());
+
+        if (xuid != null) {
+            map.put("xuid2", xuid.toString());
+        }
+
         map.put("mojToken", mojToken);
         map.put("mojTokenExpire", mojTokenExpire.toString());
         return map;
diff --git a/src/main/java/dev/figboot/olauncher/auth/ui/DeviceCodeDialog.java b/src/main/java/dev/figboot/olauncher/auth/ui/DeviceCodeDialog.java
new file mode 100644
index 0000000..580a4bd
--- /dev/null
+++ b/src/main/java/dev/figboot/olauncher/auth/ui/DeviceCodeDialog.java
@@ -0,0 +1,10 @@
+package dev.figboot.olauncher.auth.ui;
+
+import javax.swing.*;
+
+public class DeviceCodeDialog extends JDialog {
+    // https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-device-code
+    public DeviceCodeDialog() {
+
+    }
+}
diff --git a/src/main/java/dev/figboot/olauncher/auth/ui/MSLoginDialog.java b/src/main/java/dev/figboot/olauncher/auth/ui/MSLoginDialog.java
index 83bd084..f44d0c5 100644
--- a/src/main/java/dev/figboot/olauncher/auth/ui/MSLoginDialog.java
+++ b/src/main/java/dev/figboot/olauncher/auth/ui/MSLoginDialog.java
@@ -25,6 +25,7 @@ import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 import javax.swing.*;
+import javax.swing.border.EmptyBorder;
 import java.awt.*;
 import java.awt.datatransfer.StringSelection;
 import java.awt.event.ActionEvent;
@@ -56,6 +57,7 @@ public class MSLoginDialog extends JDialog {
 
         setDefaultCloseOperation(DO_NOTHING_ON_CLOSE);
 
+        lblInfo.setBorder(new EmptyBorder(5, 5, 5, 5));
         btnCancel.addActionListener(this::btnCancelClick);
         btnCopy.addActionListener(this::btnCopyLink);
 
@@ -77,9 +79,10 @@ public class MSLoginDialog extends JDialog {
         });
 
         try {
-            loginServer = new MSLoginServer(InetAddress.getByName("127.0.0.1"), OLauncherConstants.REDIR_URI_PORT, this::handleComplete);
+            loginServer = new MSLoginServer(InetAddress.getLoopbackAddress(), OLauncherConstants.REDIR_URI_PORT, this::handleComplete);
             loginServer.listen();
-            OperatingSystem.openLink(URI.create(link = String.format("https://login.live.com/oauth20_authorize.srf?client_id=%s&response_type=code&redirect_uri=%s&scope=%s",
+            OperatingSystem.openLink(URI.create(link = String.format("https://login.live.com/oauth20_authorize.srf" +
+                            "?client_id=%s&response_type=code&redirect_uri=%s&scope=%s&prompt=select_account",
                     URLEncoder.encode(OLauncherConstants.APP_CLIENT_ID.toString(), "UTF-8"),
                     URLEncoder.encode(OLauncherConstants.REDIR_URI, "UTF-8"),
                     URLEncoder.encode(OLauncherConstants.APP_SCOPES, "UTF-8")
@@ -97,6 +100,8 @@ public class MSLoginDialog extends JDialog {
     }
 
     private void initInterface() {
+        setTitle("OLauncher Login");
+        setResizable(false);
         setModal(true);
         setModalityType(ModalityType.APPLICATION_MODAL);
         setLayout(new BorderLayout());
diff --git a/src/main/java/dev/figboot/olauncher/launcher/update/ReleaseInfo.java b/src/main/java/dev/figboot/olauncher/launcher/update/ReleaseInfo.java
index c0a6ba0..74026c7 100644
--- a/src/main/java/dev/figboot/olauncher/launcher/update/ReleaseInfo.java
+++ b/src/main/java/dev/figboot/olauncher/launcher/update/ReleaseInfo.java
@@ -83,7 +83,7 @@ public class ReleaseInfo {
         }
 
         public boolean isRedistJar() {
-            return getState().equals("uploaded") && getContentType().equals("application/x-java-archive") && getName().endsWith("-redist.jar");
+            return getState().equals("uploaded") && (getContentType().equals("application/x-java-archive") || getContentType().equals("application/java-archive")) && getName().endsWith("-redist.jar");
         }
     }
 }
diff --git a/src/main/java/net/minecraft/launcher/ui/popups/login/AuthErrorForm.java b/src/main/java/net/minecraft/launcher/ui/popups/login/AuthErrorForm.java
index 7731ed5..4bfa493 100644
--- a/src/main/java/net/minecraft/launcher/ui/popups/login/AuthErrorForm.java
+++ b/src/main/java/net/minecraft/launcher/ui/popups/login/AuthErrorForm.java
@@ -25,7 +25,7 @@ public class AuthErrorForm extends JPanel {
     }
 
     protected void createInterface() {
-        this.setBorder(new EmptyBorder(0, 0, 15, 0));
+        this.setBorder(new EmptyBorder(0, 0, 5, 0)); // olauncher: shrink
         this.errorLabel.setFont(this.errorLabel.getFont().deriveFont(1));
         this.add(this.errorLabel);
     }
diff --git a/src/main/java/net/minecraft/launcher/ui/popups/login/ExistingUserListForm.java b/src/main/java/net/minecraft/launcher/ui/popups/login/ExistingUserListForm.java
index 7a9da9f..c1f6d05 100644
--- a/src/main/java/net/minecraft/launcher/ui/popups/login/ExistingUserListForm.java
+++ b/src/main/java/net/minecraft/launcher/ui/popups/login/ExistingUserListForm.java
@@ -90,7 +90,9 @@ public class ExistingUserListForm extends JPanel implements ActionListener {
         var1.gridwidth = 2;
         this.add(Box.createVerticalStrut(5), var1);
         this.add(new JLabel("Alternatively, log in with a new account below:"), var1);
-        this.add(new Separator(), var1);
+
+        // olauncher: remove separator
+        // this.add(new Separator(), var1);
     }
 
     public void actionPerformed(ActionEvent var1) {
diff --git a/src/main/java/net/minecraft/launcher/ui/popups/login/LogInForm.java b/src/main/java/net/minecraft/launcher/ui/popups/login/LogInForm.java
index 31f1846..0063b64 100644
--- a/src/main/java/net/minecraft/launcher/ui/popups/login/LogInForm.java
+++ b/src/main/java/net/minecraft/launcher/ui/popups/login/LogInForm.java
@@ -59,8 +59,9 @@ public class LogInForm extends JPanel implements ActionListener {
         var1.gridy = -1;
         var1.weightx = 1.0D;
         this.add(Box.createGlue());
+        Font var3 = getFont().deriveFont(1);
+        /*
         JLabel var2 = new JLabel("Email Address or Username:");
-        Font var3 = var2.getFont().deriveFont(1);
         Font var4 = var2.getFont().deriveFont((float)var3.getSize() - 2.0F);
         var2.setFont(var3);
         this.add(var2, var1);
@@ -90,6 +91,7 @@ public class LogInForm extends JPanel implements ActionListener {
             }
         });
         this.add(var7, var1);
+         */
         this.createUserDropdownPanel(var3);
         this.add(this.userDropdownPanel, var1);
         this.add(Box.createVerticalStrut(10), var1);
diff --git a/src/main/java/net/minecraft/launcher/ui/popups/login/LogInPopup.java b/src/main/java/net/minecraft/launcher/ui/popups/login/LogInPopup.java
index 0c66c11..d60526e 100644
--- a/src/main/java/net/minecraft/launcher/ui/popups/login/LogInPopup.java
+++ b/src/main/java/net/minecraft/launcher/ui/popups/login/LogInPopup.java
@@ -30,7 +30,8 @@ public class LogInPopup extends JPanel implements ActionListener {
     private final LogInForm logInForm;
     private final JButton loginButton = new JButton("Log In");
     private final JButton registerButton = new JButton("Register");
-    private final JButton microsoftLoginButton = new JButton("Log In with Microsoft"); // olauncher - Add MS login button
+    private final JButton microsoftLoginButton = new JButton("Microsoft Login"); // olauncher - Add MS login button
+    private final JButton microsoftDeviceLoginButton = new JButton("Microsoft Login (Device Code)"); // olauncher - Add MS login button
     private final JProgressBar progressBar = new JProgressBar();
 
     public LogInPopup(Launcher var1, LogInPopup.Callback var2) {
@@ -58,7 +59,7 @@ public class LogInPopup extends JPanel implements ActionListener {
                 JPanel var4 = new JPanel();
                 var4.add(var3);
                 this.add(var4);
-                this.add(Box.createVerticalStrut(10));
+                //this.add(Box.createVerticalStrut(10));
             }
         } catch (IOException var5) {
             var5.printStackTrace();
@@ -70,12 +71,15 @@ public class LogInPopup extends JPanel implements ActionListener {
 
         this.add(this.errorForm);
         this.add(this.logInForm);
-        this.add(Box.createVerticalStrut(15));
+        //this.add(Box.createVerticalStrut(15));
         JPanel var6 = new JPanel();
+        var6.setLayout(new GridLayout(2, 1, 0, 5));
         /*var6.setLayout(new GridLayout(1, 2, 10, 0));
         var6.add(this.registerButton);
         var6.add(this.loginButton);*/
         // olauncher - Add MS login button
+        // olauncher - remove stuff
+        /*
         var6.setLayout(new GridBagLayout());
         GridBagConstraints regConstraints = new GridBagConstraints(),
                            loginConstraints = new GridBagConstraints(),
@@ -95,7 +99,13 @@ public class LogInPopup extends JPanel implements ActionListener {
         msLoginConstraints.gridy = 1;
         msLoginConstraints.gridwidth = 2;
         msLoginConstraints.insets = new Insets(5, 0, 0, 0);
-        var6.add(microsoftLoginButton, msLoginConstraints);
+        var6.add(microsoftLoginButton, msLoginConstraints); */
+
+        microsoftLoginButton.setAlignmentX(Component.CENTER_ALIGNMENT);
+        microsoftDeviceLoginButton.setAlignmentX(Component.CENTER_ALIGNMENT);
+
+        var6.add(microsoftLoginButton);
+        var6.add(microsoftDeviceLoginButton);
 
         this.add(var6);
         this.progressBar.setIndeterminate(true);
diff --git a/src/main/resources/log4j2.xml b/src/main/resources/log4j2.xml
index 720caf7..0e7c4dd 100644
--- a/src/main/resources/log4j2.xml
+++ b/src/main/resources/log4j2.xml
@@ -13,7 +13,7 @@
         </Async>
     </Appenders>
     <Loggers>
-        <Root level="info">
+        <Root level="debug">
             <AppenderRef ref="Async"/>
         </Root>
     </Loggers>
-- 
2.43.0

