From d85366718b1d7e351770b69bf02e084102aa93a3 Mon Sep 17 00:00:00 2001
From: vops <admin@vops.cc>
Date: Tue, 11 Jun 2024 18:25:41 -0400
Subject: [PATCH] Implement-demo-profiles


diff --git a/src/main/java/dev/figboot/olauncher/auth/MCProfileInfo.java b/src/main/java/dev/figboot/olauncher/auth/MCProfileInfo.java
index c4f4b3f..9446b71 100644
--- a/src/main/java/dev/figboot/olauncher/auth/MCProfileInfo.java
+++ b/src/main/java/dev/figboot/olauncher/auth/MCProfileInfo.java
@@ -1,7 +1,9 @@
 package dev.figboot.olauncher.auth;
 
+import lombok.AllArgsConstructor;
 import lombok.Getter;
 
+import java.util.ArrayList;
 import java.util.List;
 import java.util.UUID;
 
@@ -9,11 +11,16 @@ import java.util.UUID;
 public class MCProfileInfo {
     private UUID id;
     private String name;
-
     private List<Skin> skins;
-
     private List<Cape> capes;
 
+    public MCProfileInfo(UUID id, String name) {
+        this.id = id;
+        this.name = name;
+        this.skins = new ArrayList<>();
+        this.capes = new ArrayList<>();
+    }
+
     @Getter
     public static class Skin {
         private UUID id;
diff --git a/src/main/java/dev/figboot/olauncher/auth/MicrosoftUserAuthentication.java b/src/main/java/dev/figboot/olauncher/auth/MicrosoftUserAuthentication.java
index 86f7f18..354a2c2 100644
--- a/src/main/java/dev/figboot/olauncher/auth/MicrosoftUserAuthentication.java
+++ b/src/main/java/dev/figboot/olauncher/auth/MicrosoftUserAuthentication.java
@@ -249,13 +249,16 @@ public class MicrosoftUserAuthentication extends HttpUserAuthentication {
     private void profileLogIn() throws AuthenticationException {
         if (this.xuid == null) findXUID();
 
-        if (!ownsGame()) {
-            throw new AuthenticationException("No Minecraft entitlement! Do you own the game?");
-        }
-
         LOGGER.debug("Getting your user profile...");
-        profileInfo = getAuthenticationService().getAuthenticated(MicrosoftAuthenticationService.ROUTE_MINECRAFT_PROFILE,
-                mojToken, MCProfileInfo.class);
+
+        boolean isDemo = !ownsGame();
+
+        if (isDemo) {
+            LOGGER.debug("No Minecraft entitlement! Utilizing demo profile...");
+            profileInfo = new MCProfileInfo(xuid, "Player");
+        } else {
+            profileInfo = getAuthenticationService().getAuthenticated(MicrosoftAuthenticationService.ROUTE_MINECRAFT_PROFILE, mojToken, MCProfileInfo.class);
+        }
 
         if (false) { // this code should be unnecessary now -figboot
         while (profileInfo.getId() == null) {
@@ -272,15 +275,24 @@ public class MicrosoftUserAuthentication extends HttpUserAuthentication {
         }
         }
 
-        GameProfile profile = new GameProfile(profileInfo.getId(), profileInfo.getName());
-        profile = sessionService.fillProfileProperties(profile, false);
+        String userId = profileInfo.getId().toString().replace("-", "");
+
+        if (isDemo) {
+            setSelectedProfile(null);
+            profiles = new GameProfile[0];
+            // demo profiles use XUID, which are usually padded with lots of zeros
+            userId = userId.replaceFirst("^0+(?!$)", "");
+        } else {
+            GameProfile profile = new GameProfile(profileInfo.getId(), profileInfo.getName());
+            profile = sessionService.fillProfileProperties(profile, false);
+            setSelectedProfile(profile);
+            getModifiableUserProperties().clear();
+            getModifiableUserProperties().putAll(profile.getProperties());
+            profiles = new GameProfile[]{profile};
+        }
 
         online = true;
-        setUserid(profileInfo.getId().toString().replace("-", ""));
-        profiles = new GameProfile[]{profile};
-        setSelectedProfile(profile);
-        getModifiableUserProperties().clear();
-        getModifiableUserProperties().putAll(profile.getProperties());
+        setUserid(userId);
         setUserType(UserType.MICROSOFT);
 
         /*try {
-- 
2.47.2

